#script to simulate test data for admixmap##simulateHaploidAlleles <- function(M,ro,x,L) {   gameteAncestry <- numeric(L)  randAnc <- runif(L)  gameteAncestry[1] <- ifelse(randAnc[1] > M, 2, 1) # M is prob pop 1.    for(locus in 2:L) {    T <- t(matrix(data=c(M + (1-M)*exp(-ro*x[locus]),  1-M - (1-M)*exp(-ro*x[locus]), M - M*exp(-ro*x[locus]), 1-M + M*exp(-ro*x[locus]) ), nrow=2, ncol=2))      gameteAncestry[locus] <- ifelse(randAnc[locus] > T[gameteAncestry[locus-1], 1], 2, 1)  }  simulateHaploidAlleles <- ifelse(runif(L) > alleleFreqs[1,gameteAncestry], 2, 1)}simulateGenotypes <- function(M1,M2, ro,x,L) {  paternalGamete <- simulateHaploidAlleles(M1,ro,x,L)  maternalGamete <- simulateHaploidAlleles(M2,ro,x,L)  simulateGenotypes <- paste(paternalGamete, ",", maternalGamete, sep="")}# Start of script#sumofintensitiesro <- 6#number of lociL <- 40#default#L <- 200 #longer chromosome#marker spacings#x <- 0.1 #evenly spaced#x <- runif(L,0.01,0.05)#default, spacings between 1 and 5 cMx <- rep(100,L)#unlinked markers#x <- rep(0.01,L) #denser markers#alleleFreqs <- matrix(c(1, 0, 0, 1), nrow=2, ncol=2)alleleFreqs <- matrix(c(0.8, 0.2, 0.2, 0.8), nrow=2, ncol=2)genotypes <- character(L)##write("outcome\n", file="regression/data/genotypes.txt", append=FALSE)outcome <- numeric(0)Mvector <- numeric(0)for(individual in 1:400) {  M1 <- rbeta(1, 8, 2)  M2 <- rbeta(1, 8, 2)# random mating  #M2 <- M1 #assortative mating  Mvector <- c(Mvector, 0.5*(M1+M2))  obs <- simulateGenotypes(M1, M2, ro, x, L)  #make some genotypes missing  for(locus in 1:L)if(runif(n=1) < 0.2)obs[locus]<-""  genotypes <- rbind(genotypes, obs)  outcome <- c(outcome, rnorm(1, mean=(50 + 25*(0.5*(M1+M2))), sd=10)) #linear regression  ##outcome <- c(outcome, rbinom(1, 1, 1 / (1+exp(-2*M2))))#binary outcome}#write outcome variable to fileoutcome <- data.frame(outcome, row.names=NULL)write.table(outcome, file="regression/data/outcome.txt", row.names=FALSE, col.names=TRUE)Mvector <- data.frama(Mvector, rownames=NULL)write.table(outcome, file="regression/data/Mvalues.txt", row.names=FALSE, col.names=TRUE)#create genotypes filegenotypes <- genotypes[-1,]id = as.character(seq(1:400))genotypes <- data.frame(id, genotypes, row.names=NULL)write.table(genotypes, file="regression/data/genotypes40unlinked.txt", row.names=FALSE)#create locus fileloci <- data.frame(as.vector(dimnames(genotypes)[[2]][-1]),                   rep(2,L),                   c(100, x[1:(L-1)]),                   #rep(x, L-1),                   row.names=NULL)dimnames(loci)[[2]] <- c("Locus", "NumAlleles", "Distance")write.table(loci, file="regression/data/loci40unlinked.txt", row.names=FALSE)#create allelefreqs filesallelefreqs <- data.frame(as.vector(loci[,1]), rep(alleleFreqs[1,1],L), rep(alleleFreqs[1,2],L))dimnames(allelefreqs)[[2]] <- c("Locus", "Pop1", "Pop2")#old format allelefreqfilewrite.table(allelefreqs, file="regression/data/allelefreqfile.txt", row.names=FALSE)priorallelefreqs <- numeric(2)locusnames <- character(0)for(locus in 1:L) {  priorallelefreqs <- rbind(priorallelefreqs, 500*alleleFreqs + 0.5)  locusnames <- c(locusnames, rep(as.vector(loci[locus,1]), 2))}#priorallelefreqfilepriorallelefreqs <- data.frame(locusnames, priorallelefreqs[-1,], row.names=NULL)write.table(priorallelefreqs, file="regression/data/priorallelefreqs.txt", row.names=FALSE)                               