<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:st1="urn:schemas-microsoft-com:office:smarttags"
xmlns="http://www.w3.org/TR/REC-html40">

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=ProgId content=FrontPage.Editor.Document>
<meta name=Generator content="Microsoft FrontPage 4.0">
<meta name=Originator content="Microsoft Word 10">
<link rel=File-List href="file:///S:/SharedFolders/genepi/Website/ADMIXMAPManual_files/filelist.xml">
<link rel=Edit-Time-Data href="file:///S:/SharedFolders/genepi/Website/ADMIXMAPManual_files/editdata.mso">
<!--[if !mso]>
<style>
v\:* {behavior:url(#default#VML);}
o\:* {behavior:url(#default#VML);}
w\:* {behavior:url(#default#VML);}
.shape {behavior:url(#default#VML);}
</style>
<![endif]-->
<title>Genetic Epidemiology Unit</title>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="date"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="City"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="PlaceType"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="place"/>
<!--[if gte mso 9]><xml>
 <o:DocumentProperties>
  <o:Author> </o:Author>
  <o:LastAuthor> </o:LastAuthor>
  <o:Revision>2</o:Revision>
  <o:TotalTime>247</o:TotalTime>
  <o:LastPrinted>2004-12-20T14:26:00Z</o:LastPrinted>
  <o:Created>2004-12-20T14:35:00Z</o:Created>
  <o:LastSaved>2004-12-20T14:35:00Z</o:LastSaved>
  <o:Pages>1</o:Pages>
  <o:Words>8530</o:Words>
  <o:Characters>48622</o:Characters>
  <o:Company> </o:Company>
  <o:Lines>405</o:Lines>
  <o:Paragraphs>114</o:Paragraphs>
  <o:CharactersWithSpaces>57038</o:CharactersWithSpaces>
  <o:Version>10.6626</o:Version>
 </o:DocumentProperties>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <w:WordDocument>
  <w:GrammarState>Clean</w:GrammarState>
  <w:BrowserLevel>MicrosoftInternetExplorer4</w:BrowserLevel>
 </w:WordDocument>
</xml><![endif]--><!--[if !mso]><object
 classid="clsid:38481807-CA0E-42D2-BF39-B33AF135CC4D" id=ieooui></object>
<style>
st1\:*{behavior:url(#ieooui) }
</style>
<![endif]-->
<style>
<!--
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{mso-style-parent:"";
	margin:0cm;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
h2
	{mso-margin-top-alt:auto;
	margin-right:0cm;
	mso-margin-bottom-alt:auto;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	mso-outline-level:2;
	font-size:18.0pt;
	font-family:"Times New Roman";
	font-weight:bold;}
a:link, span.MsoHyperlink
	{color:#214772;
	text-decoration:underline;
	text-underline:single;}
a:visited, span.MsoHyperlinkFollowed
	{color:maroon;
	text-decoration:underline;
	text-underline:single;}
p
	{mso-margin-top-alt:auto;
	margin-right:0cm;
	mso-margin-bottom-alt:auto;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
p.stylebefore5ptafter5pt, li.stylebefore5ptafter5pt, div.stylebefore5ptafter5pt
	{mso-style-name:stylebefore5ptafter5pt;
	mso-margin-top-alt:auto;
	margin-right:0cm;
	mso-margin-bottom-alt:auto;
	margin-left:0cm;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
span.grame
	{mso-style-name:grame;}
span.GramE
	{mso-style-name:"";
	mso-gram-e:yes;}
@page Section1
	{size:612.0pt 792.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;
	mso-header-margin:35.4pt;
	mso-footer-margin:35.4pt;
	mso-paper-source:0;}
div.Section1
	{page:Section1;}
 /* List Definitions */
 @list l0
	{mso-list-id:534929214;
	mso-list-template-ids:892252634;}
@list l1
	{mso-list-id:772169985;
	mso-list-template-ids:1745929786;}
@list l2
	{mso-list-id:1958752687;
	mso-list-template-ids:-2017052014;}
ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>
<!--[if gte mso 10]>
<style>
 /* Style Definitions */
 table.MsoNormalTable
	{mso-style-name:"Table Normal";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:0;
	mso-style-noshow:yes;
	mso-style-parent:"";
	mso-padding-alt:0cm 5.4pt 0cm 5.4pt;
	mso-para-margin:0cm;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";}
</style>
<![endif]-->
</head>

<body lang=EN-US link="#214772" vlink=maroon style='tab-interval:36.0pt'
alink="#003300">

<div class=Section1>

<table class=MsoNormalTable border=0 cellpadding=0 width="100%"
 style='width:100.0%;mso-cellspacing:1.5pt;mso-padding-alt:3.55pt 3.55pt 3.55pt 3.55pt'
 height="7%" bordercolorlight="#FFFFFF" bordercolordark="#FFFFFF">
 <tr style='mso-yfti-irow:0;mso-yfti-lastrow:yes;height:41.35pt'>
  <td width="16%" style='width:16.0%;padding:3.55pt 3.55pt 3.55pt 3.55pt;
  height:41.35pt'>
  <p class=MsoNormal>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <img border="0" src="../../../../Documents/Website/images/dna4.jpg" width="93" height="128"></p>
  </td>
  <td width="69%" style='width:69.0%;padding:3.55pt 3.55pt 3.55pt 3.55pt;
  height:41.35pt'>
  <p><a name="Genetic_Epidemiology_Unit"><span style='font-size:36.0pt;
  font-family:Arial;color:#214772'>Genetic Epidemiology Unit</span></a></p>
  <p><st1:place><st1:PlaceType><i><span style='font-size:24.0pt;font-family:
    Arial;color:#214772'>University</span></i></st1:PlaceType><i><span
   style='font-size:24.0pt;font-family:Arial;color:#214772'> </span></i><st1:PlaceType><i><span
    style='font-size:24.0pt;font-family:Arial;color:#214772'>College</span></i></st1:PlaceType></st1:place><i><span
  style='font-size:24.0pt;font-family:Arial;color:#214772'> </span></i><st1:City><st1:place><i><span
    style='font-size:24.0pt;font-family:Arial;color:#214772'>Dublin</span></i></st1:place></st1:City></p>
  </td>
 </tr>
</table>

<p class=MsoNormal align=center style='text-align:center'><span
style='display:none;mso-hide:all'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span style='display:none;mso-hide:all'></o:p></span></p>

<table class=MsoNormalTable border=0 cellpadding=0 width="98%"
 style='width:98.0%;mso-cellspacing:1.5pt' bordercolorlight="#FFFFFF"
 bordercolordark="#FFFFFF" height=334>
 <tr style='mso-yfti-irow:0;mso-yfti-lastrow:yes;height:235.25pt'>
  <td width="87%" style='width:87.0%;padding:.75pt .75pt .75pt .75pt;
  height:235.25pt'>
  <p style='margin-left:7.15pt'><b><span style='font-size:13.5pt;font-family:
  Arial'>&nbsp; </span></b><!-- Start of Menu --><!-- End of Menu --></p>
  <div>
  <p><span style='font-size:13.5pt;font-family:Arial;color:#214772'><b>ADMIXMAP</b> -
  a program to model admixture using marker genotype data</span></p>
  <p><a href="#Description_of_program:"><span style='font-family:Arial'>Description
  of ADMIXMAP</span></a></p>
  <p><a href="#Applications_of_the_program:"><span style='font-family:Arial'>Applications
  of ADMIXMAP</span></a></p>
  <p><a href="#Running_the_program:"><span style='font-family:Arial'>Running ADMIXMAP</span></a></p>
  <p class=stylebefore5ptafter5pt><a href="#Options_specified_by_the_user:"><span
  style='font-family:Arial'>Options specified by the user</span></a></p>
  <p class=stylebefore5ptafter5pt><a href="#Input File Options"><span
  style='font-family:Arial'>Input file</span><span style="font-family: Arial">
  options</span></a></p>
  <p class=stylebefore5ptafter5pt><a href="#Output_files:"><span
  style='font-family:Arial'>Output files</span></a> </p>
  <p class=stylebefore5ptafter5pt><span style="font-family: Arial"><a href="#Interpretation of output from the program">Interpretation
  of output from the program</a></span> </p>
  <p><a href="#Contact:"><span style='font-family:Arial'>Contact</span></a></p>
  <div>
  <div class=MsoNormal align=center style='text-align:center'>
  <hr size=2 width="100%" align=center>
  </div>
  </div>
  <p><a name="Description_of_program:"><span style='font-size:13.5pt;
  font-family:Arial;color:#214772'>Description of program:</span></a></p>
  <p><span style='font-family:Arial'>ADMIXMAP is designed to analyse datasets
  that consist of trait measurements and genotype data on a sample of
  individuals from an admixed or stratified population. Although the name of
  the program reflects its origins as a program designed for admixture mapping,
  it has wider uses, especially in genetic association studies. The study
  design can be a cross-sectional survey of a quantitative trait or binary
  outcome, a case-control study or a cohort study. For admixture to be modelled
  efficiently, at least some of the loci typed should be
  &quot;ancestry-informative markers&quot;: markers chosen to have large allele
  frequency differentials between the ancestral subpopulations that underwent
  admixture. The program can deal with any number of ancestral subpopulations
  and any number of linked marker loci. In its present version, the program
  handles only data from samples of unrelated individuals.</span></p>
  <p><span style='font-family:Arial'>The program is written in C++, and is freely
  available with source code under a GPL. <span
  style='mso-spacerun:yes'> </span>Offers to help with development of the
  program are welcome. <span style='mso-spacerun:yes'> </span>The current
  version runs only on a single processor, and computation time is a serious
  limitation on large datasets.<span style='mso-spacerun:yes'>  </span>We are
  developing a parallel version that will be able to run on a computing
  cluster, using the MPICH implementation of the MPI message-passing standard<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>The program is based on a hybrid of
  Bayesian and classical approaches. A Bayesian full probability model is
  specified, assigning vague prior distributions to parameters for the
  distribution of admixture in the population and the stochastic variation of
  ancestry along hybrid chromosomes. The posterior distribution of all
  unobserved variables given the observed genotype and trait <span class=GramE>data,</span>
  is generated by Markov chain </span><st1:place><span style='font-family:Arial'>Monte
   Carlo</span></st1:place><span style='font-family:Arial'> simulation. These
  unobserved variables include the ancestry at each locus and the
  ancestry-specific allele frequencies at each locus. For a description of the
  theory underlying this approach, see the following papers:-</span> </p>
  <p><span style='font-family:Arial'>McKeigue, P.M., Carpenter, J., Parra,
  E.J., Shriver, <span class=GramE><span class=grame>M.D..</span></span>
  <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=11246470">
  Estimation of admixture and detection of linkage in admixed populations by a
  Bayesian approach: application to African-American populations</a>. <i>Annals of
  Human Genetics</i> 2000<span class=GramE><span class=grame>;<b>64</b></span></span>:
  171-86.</span> </p>
  <p><span style='font-family:Arial'>Hoggart, C.J., Parra, E.J., Shriver, M.D.,
  Bonilla, C., Kittles, R.A., Clayton, D.G. and McKeigue, P.M. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12817591"> Control of
  confounding of genetic associations in stratified populations</a>. <i>Am J Hum
  Genet</i>. 2003; <b>72</b>:1492-1504.</span></p>
  <p><span style='font-family:Arial'>Hoggart, C.J., Shriver, M.D., Kittles,
  R.A., Clayton, D.G. and McKeigue, P.M. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15088268"> Design and analysis of admixture
  mapping studies</a>.<i>Am J Hum Genet</i>. 2004; <b>74</b>:965-978.<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>McKeigue, PM; <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15540159"> Prospects for admixture
  mapping of complex traits</a><i style='mso-bidi-font-style:normal'>. Am J Hum
  Genet</i> 2005; in press</span></p>
  <p>&nbsp;</p>
  <p><a name="Applications_of_the_program:"><span style='font-size:13.5pt;
  font-family:Arial;color:#214772'>Applications of the program:</span></a></p>
  <p><i><span style='font-family:Arial'>1. Modelling the dependence of a
  disease or quantitative trait upon individual admixture</span></i><span
  style='font-family:Arial'><br>
  For a binary trait, such as presence of disease, the program fits a logistic
  regression model of the trait upon individual admixture, mean of admixture
  proportions of both parents. For a continuous trait, such as skin
  pigmentation, the program fits a linear regression model of the trait value
  on individual admixture. Covariates such as age, sex and socioeconomic status
  can be included in the regression model. The program output includes
  posterior means and 95% credible intervals for the regression coefficients.
  Alternatively, the program can be used to test a null hypothesis of no
  association of disease risk or trait level effect with individual admixture
  as described below.</span> </p>
  <p><i><span style='font-family:Arial'>2. Controlling for confounding of
  genetic associations in stratified populations</span></i><span
  style='font-family:Arial'><br>
  <span class=grame>For</span> more details of this application, see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12817591"> Hoggart et
  al (2003)</a>. The program calculates a score test for association of the disease
  or trait with alleles or haplotypes at each locus, adjusting for individual
  admixture and other covariates in a regression model. Where there is evidence
  for association of a trait with individual admixture, the posterior
  distribution of the regression coefficient can be estimated in a further
  analysis.&nbsp; For this application, the dataset should include at least 30
  markers informative for ancestry.&nbsp; </span></p>
  <p><i><span style='font-family:Arial'>3. Admixture mapping: localizing genes
  that underlie ethnic differences in disease risk. </span></i><span
  style='font-family:Arial'><br>
  For more details of this application, see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=15088268"> Hoggart et al
  (2004)</a>. Where
  differences in disease risk have a genetic basis, testing for association of
  the disease with locus ancestry by conditioning on parental admixture can
  localize genes underlying these differences. This approach is an extension of
  the principles underlying linkage analysis of an experimental cross. To
  exploit the full power of admixture mapping, 1000 or more markers informative
  for ancestry across the genome are required.<span
  style='mso-spacerun:yes'>    </span></span></p>
  <p><i><span style='font-family:Arial'>4. Detecting population stratification
  and identifying admixed individuals<br>
  </span></i><span class=grame><span style='font-family:Arial'>Where</span></span><span
  style='font-family:Arial'> no information about the demographic background of
  the population under study is available, ADMIXMAP can be used to test for
  population stratification, to determine how many subpopulations are required
  to model this stratification, and to identify admixed individuals.&nbsp; This
  is useful when assembling panels of unadmixed individuals to be used for
  estimating allele frequencies.&nbsp; We emphasize that when the program is
  run without supplying prior information about allele frequencies in each
  subpopulation, the subpopulations are not identifiable in the model.<span
  style='mso-spacerun:yes'>  </span>Thus inference should be based only <span
  class=GramE><span class=grame>on<span style='mso-spacerun:yes'>  </span>the</span></span>
  posterior distribution of variables that are unaffected by permuting the
  labels of the subpopulations.&nbsp; </span></p>
  <p><i><span style='font-family:Arial'>5. Testing for associations of a trait
  with haplotypes and estimating haplotype frequencies from a sample of
  unrelated individuals<br>
  </span></i><span class=grame><span style='font-family:Arial'>Where</span></span><span
  style='font-family:Arial'> two or more loci in the same gene have been typed,
  ADMIXMAP will model the unobserved haplotypes, conditional on the observed
  unordered genotypes.&nbsp;&nbsp; Score tests for association of haplotypes
  with the trait can be obtained, and samples from the posterior distribution
  of haplotype frequencies can be obtained.<span style='mso-spacerun:yes'> 
  </span>This application of the program is not limited to admixed or
  stratified populations: for a population that is not stratified, the user can
  simply specify the option <i>populations</i>=1</span></p>
  <p><span style='font-family:Arial'>For each of these applications, score
  tests of the appropriate null hypotheses are built into the program.</span> </p>
  <p><a name="Modelling_admixture_and_trait_values:"><span style='font-size:
  13.5pt;font-family:Arial;color:#214772'>Modelling admixture and trait values:</span></a></p>
  <p><span style='font-family:Arial'>Any run of two or more loci for which the
  distance between loci is specified as zero is modelled as a single
  &quot;compound locus&quot;.<span style='mso-spacerun:yes'> </span>Thus if L
  &quot;simple loci&quot; (SNPs, insertion/deletion polymorphisms or microsatellites)
  have been typed, and three of these simple loci are in the same gene, the
  model will have L - 2 compound loci.<span style='mso-spacerun:yes'> </span>The program assumes that on any gamete, the ancestry state is the same
  at all loci within a compound locus. The program allows for allelic
  association within any compound locus that contains two or more simple loci,
  and models the unobserved haplotypes at this compound locus.</span> </p>
  <p><span style='font-family:Arial'>For each parent of each individual,
  admixture proportions are defined by a vector with <i style='mso-bidi-font-style:
  normal'>K</i> co-ordinates, where <i style='mso-bidi-font-style:normal'>K</i>
  is the number of ancestral subpopulations that contributed to the admixed
  population under study.<span style='mso-spacerun:yes'> </span>For instance,
  in a <st1:place>Caribbean</st1:place> population it may be possible to model
  the gene pool of the admixed population as a mixture of three subpopulations:
  African, European and Native American.<span style='mso-spacerun:yes'> </span>The model of admixture is described by the following hierarchy:-</span></p>
  <p class=stylebefore5ptafter5pt><span style='font-family:Arial'>1. The
  population distribution from which the parental admixture proportions are
  drawn is modelled by a Dirichlet distribution with parameter vector of length
  <i style='mso-bidi-font-style:normal'>K</i>. </span></p>
  <p class=stylebefore5ptafter5pt><span style='font-family:Arial'>2. The allele
  or haplotype frequencies at each compound locus are modelled by a Dirichlet
  distribution, with prior parameters specified by the user.</span> </p>
  <p class=stylebefore5ptafter5pt><span style='font-family:Arial'>3. Locus
  ancestry is modelled by a multinomial distribution, with cell probabilities
  specified by the admixture of both parents.</span> </p>
  <p class=stylebefore5ptafter5pt><span style='font-family:Arial'>4. The
  probabilities of observing each allele or haplotype at each locus on each
  gamete, given the ancestry of the gamete at that locus, are modelled by a
  multinomial distribution, with parameters given by the ancestry-specific
  allele (or haplotype) frequencies.</span> </p>
  <p><span style='font-family:Arial'>5. The stochastic variation of ancestral
  states along the chromosomes transmitted from each parent is modelled as a
  mixture of independent Poisson arrival processes with <span class=GramE>intensities
  </span></span><span style='font-family:Symbol;mso-ascii-font-family:Arial;
  mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;
  mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:
  Symbol'>a</span></span><span style='font-family:Arial'>, </span><span
  style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
  Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>b</span></span><span
  style='font-family:Arial'>, </span><span style='font-family:Symbol;
  mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
  Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
  style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>g</span></span><span
  style='font-family:Arial'> per Morgan (for three-way admixture). For given
  values of parental admixture, it is only necessary to specify a single
  parameter </span><span style='font-family:Symbol;mso-ascii-font-family:Arial;
  mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;
  mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:
  Symbol'>r</span></span><span style='font-family:Arial'> for the sum of
  intensities: </span><span style='font-family:Symbol;mso-ascii-font-family:
  Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:
  symbol;mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;
  mso-symbol-font-family:Symbol'>r</span></span><span style='font-family:Arial'>=
  </span><span style='font-family:Symbol;mso-ascii-font-family:Arial;
  mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;
  mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:
  Symbol'>a</span></span><span style='font-family:Arial'> + </span><span
  style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
  Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>b</span></span><span
  style='font-family:Arial'> <span class=GramE>+ </span></span><span
  style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
  Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>g</span></span><span
  style='font-family:Arial'>.</span> </p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>If an outcome variable is supplied<span
  class=GramE>,<span style='mso-spacerun:yes'>  </span>ADMIXMAP</span> fits a
  regression model (logistic regression for a binary trait, linear regression
  for a quantitative trait) with individual admixture proportions and any
  covariates supplied by the user as explanatory variables.&nbsp; </span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'>&nbsp;</p>
  <p><a name="Modelling_allele/haplotype_frequencies:"><span style='font-size:
  13.5pt;font-family:Arial;color:#214772'>Modelling allele/haplotype
  frequencies:</span></a></p>
  <p><span style='font-family:Arial'>The program can be run with
  ancestry-specific allele / haplotype frequencies specified either as fixed or
  as random variables.&nbsp; If one of the two options <i>allelefreqfile</i> or
  <i style='mso-bidi-font-style:normal'>fixedallelefreqs</i> is specified, the
  allele frequencies are specified as fixed at the values supplied.<span
  style='mso-spacerun:yes'>  </span>If <span class=GramE>option <i>populations</i>
  is</span> specified, the allele frequencies are specified as random variables
  with reference (uninformative) prior distributions.&nbsp;&nbsp; If option <i>priorallelefreqsfile</i>
  is specified, the allele frequencies are specified as random variables with a
  prior distribution given by the values in this file.<span
  style='mso-spacerun:yes'>   </span>This option is used where allele
  frequencies have been estimated from samples of unadmixed modern descendants
  of the ancestral subpopulations that contributed to the admixed population
  under study.<span style='mso-spacerun:yes'>  </span>For instance, in a study
  of a population of mixed European and west African ancestry, allele
  frequencies at some or all of the loci typed may have been estimated in
  samples from modern unadmixed west African and European populations. The
  program will use this information to estimate the ancestry-specific allele
  frequencies from the unadmixed and admixed population samples simultaneously,
  allowing for sampling error.</span> </p>
  <p><span style='font-family:Arial'>If no information about allele frequencies
  in the ancestral subpopulations is provided, the ancestry-specific allele
  frequencies are estimated only from the admixed population under study. If no
  information about allele frequencies is provided at any locus, the
  subpopulations are not identifiable in the model. This does not matter when
  the program is used only to control for confounding by hidden population
  stratification, as described in <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=12817591"> Hoggart et al.
  (2003)</a>.</span> <span
  style='mso-spacerun:yes'>  </span></p>
  <p><span style='font-family:Arial'>The file <i>priorallelefreqfile</i>
  specifies the parameters of a Dirichlet prior distribution for the allele
  frequencies at each locus in each subpopulation.<span
  style='mso-spacerun:yes'>   </span>Where the alleles or haplotypes have been
  counted directly in samples from unadmixed modern descendants, these
  parameter values should be specified by adding 0.5 to the observed counts of
  each allele or haplotype in each subpopulation.<span
  style='mso-spacerun:yes'>   </span>These parameter values specify the
  Dirichlet posterior distribution that we would obtain by combining a
  reference prior with the observed counts.&nbsp; Using this as a prior
  distribution when analysing data from the admixed population is equivalent to
  estimating the allele frequencies simultaneously from the admixed and
  unadmixed population samples, with a reference prior.</span> </p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>For compound loci, where haplotype frequencies have
  been estimated from unordered genotypes rather than by counting phase-known
  gametes, the user cannot specify the prior distribution simply by adding 0.5
  to the observed counts of each haplotype in a sample of unadmixed modern
  descendants, because the counts have not been observed.&nbsp; Instead, we can
  compute the posterior distribution of haplotype frequencies in the unadmixed
  population, given a reference prior and the observed unordered genotype data.<span
  style='mso-spacerun:yes'>  </span>In accordance with the principles of
  Bayesian inference, we can then use this posterior distribution to specify the
  prior distribution when modelling data from the admixed population.<span
  style='mso-spacerun:yes'>  </span>To simplify the computation, we generate a
  large sample from the posterior distribution of haplotype frequencies in the
  unadmixed population and calculate the parameters of the Dirichlet
  distribution that most closely approximates this posterior distribution.<span
  style='mso-spacerun:yes'>  </span>The parameters of this Dirichlet
  distribution are then entered in file <i>priorallelefreqfile</i>.<span
  style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>This can be implemented by running ADMIXMAP with genotype
  data from each unadmixed population sample, specifying options <i>populations</i>=1
  and <i>allelefreqoutputfile</i> to sample the posterior distribution of the
  haplotype frequencies. &nbsp;The parameters of the Dirichlet distribution that
  most closely approximates this posterior distribution can then be computed,
  and substituted into the file <i>priorallelefreq </i>as input to the program
  when modelling data from the admixed population.<span
  style='mso-spacerun:yes'>  </span>This is straightforward: for each locus and
  each subpopulation, the posterior expectations and the posterior covariance
  matrix of the allele frequencies are evaluated using the samples from the
  posterior distribution in <i>allelefreqoutputfile</i>.<span
  style='mso-spacerun:yes'>  </span>The Dirichlet parameters </span><b><span
  style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
  Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>a</span></span></b><b><sub><span
  style='font-family:Arial'>i</span></sub></b><b><span style='font-family:Arial'>
  </span></b><span style='font-family:Arial;mso-bidi-font-weight:bold'>that
  approximate the posterior distribution </span><span style='font-family:Arial'>are
  then computed by equating the posterior expectations of the allele
  frequencies to the ratios </span><b><span style='font-family:Symbol;
  mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
  Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
  style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>a</span></span></b><b><sub><span
  style='font-family:Arial'>i</span></sub></b><b><span style='font-family:Arial'>
  / </span></b><b><span style='font-family:Symbol;mso-ascii-font-family:Arial;
  mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;
  mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:
  Symbol'>a</span></span></b><b><span style='font-family:Symbol;mso-ascii-font-family:
  Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:
  symbol;mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;
  mso-symbol-font-family:Symbol'>S</span></span></b><b><sub><span
  style='font-family:Arial'>i</span></sub></b><span style='font-family:Arial'>,
  and equating the determinant of the posterior covariance matrix to the
  determinant of the covariance matrix of the Dirichlet distribution.<b><span
  style='mso-spacerun:yes'>  </span></b>If ADMIXMAP is run with the option <i
  style='mso-bidi-font-style:normal'>allelefreqoutputfile</i> and the R script <i
  style='mso-bidi-font-style:normal'>AdmixmapOutput.R</i> is run to process the
  output files, the Dirichlet parameters will be computed and written to a file
  in the correct format for use in subsequent analyses as <i style='mso-bidi-font-style:
  normal'>priorallelefreqfile</i>.<span style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p><span style='font-family:Arial'>With options <i>populations</i>, <i>allelefreqfile</i>
  or <i>priorallelefreqfile</i>, the program fits a model in which the allele
  frequencies in modern unadmixed descendants of the ancestral subpopulations
  are identical to the corresponding ancestry-specific allele frequencies in
  the admixed population under study.<span style='mso-spacerun:yes'> 
  </span>The option <i>dispersiontestfile</i> will generate a diagnostic test
  of this assumption.<span style='mso-spacerun:yes'>  </span></span></p>
  <p><span style='font-family:Arial'>With option <i>historicallelefreqfile</i>,
  the program fits a more general model in which there is dispersion of allele
  frequencies between the unadmixed and admixed populations.</span> </p>
  <p>&nbsp;<a name="Inference:"><span style='font-size:13.5pt;font-family:Arial;
  color:#214772'>Inference:</span></a></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>There are various approaches to statistical
  inference and hypothesis testing using the ADMIXMAP program:-<span
  style='mso-spacerun:yes'>  </span></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>(1)</span><span style='font-size:7.0pt'> </span><span
  class=GramE><span style='font-family:Arial'>a</span></span><span
  style='font-family:Arial'> model based on the null hypothesis can be fitted,
  and this null hypothesis can be tested against alternatives with a score test
  computed by averaging over the posterior distribution of the missing
  data.<span style='mso-spacerun:yes'>  </span>For a description of the theory
  underlying this approach, see Hoggart et al. (2003). Several score tests are
  built into the program, and are described below.<span
  style='mso-spacerun:yes'>  </span>Additional score tests can be constructed
  by the user.&nbsp; </span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>(2)</span><span style='font-size:7.0pt'> </span><span
  class=GramE><span style='font-family:Arial'>the</span></span><span
  style='font-family:Arial'> effect under study can be included in the regression
  model, so that the posterior distribution of the effect parameter is
  estimated.<span style='mso-spacerun:yes'>  </span>In large samples, the
  posterior mean and 95% credible interval for the effect parameter are
  asymptotically equivalent to the maximum likelihood estimate and 95%
  confidence interval that would be obtained by classical methods.&nbsp; <o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>(3) the log-likelihood function for the effect of a
  parameter can be computed: not yet implemented <o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>(4) <span class=GramE>the</span> marginal
  likelihood of the model can be evaluated using Chib’s algorithm.<span
  style='mso-spacerun:yes'>  </span>At present this is implemented only for an
  individual-level model, in which there is no inference on population
  parameters and no outcome variable.<span style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>In addition to these methods for formal inference, model
  diagnostics, based on the posterior predictive check probability, are
  provided to detect population stratification not accounted for in the model, and
  lack of fit of the allele frequencies to the specified model. <o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'>&nbsp;</p>
  <p><a name="Comparison_with_other_programs_for_model"><span style='font-size:
  13.5pt;font-family:Arial;color:#214772'>Comparison with other programs for
  modelling admixture:</span></a></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>The program STRUCTURE (available from <a
  href="http://pritch.bsd.uchicago.edu/">http://pritch.bsd.uchicago.edu/</a>) fits
  a similar hierarchical model for population admixture, given genotype data on
  admixed and unadmixed individuals, if you specify the &quot;popalphas&quot;
  option (see documentation for this program at
  http://pritch.bsd.uchicago.edu/software/readme_structure2.pdf).<span
  style='mso-spacerun:yes'>  </span></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial'>The main differences between ADMIXMAP and STRUCTURE
  are:-<span style='mso-spacerun:yes'>  </span></span></p>
  <ol start=1 type=1>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l2 level1 lfo1;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>STRUCTURE does not model the dependence of the
       outcome variable on individual admixture and thus cannot adjust for the
       effect of individual admixture on the outcome variable.</span> </li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l2 level1 lfo1;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>STRUCTURE does not allow the user to supply
       prior distributions for the allele frequencies.&nbsp; To use allele
       frequency data from unadmixed individuals in STRUCTURE, individuals
       sampled from unadmixed and admixed populations have to be included in
       the same model.<span style='mso-spacerun:yes'>  </span>This is not
       recommended, either with STRUCTURE or ADMIXMAP, because the model
       assumes a unimodal distribution of individual admixture values in the
       population.&nbsp; If samples from both unadmixed and admixed populations
       are included in the same model, the distribution of individual admixture
       values will generally not be unimodal, and the fit of the model will be
       poor.<span style='mso-spacerun:yes'>  </span></span></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l2 level1 lfo1;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>STRUCTURE does not allow for allelic
       association (other than that generated by admixture), and is therefore
       unsuitable for analysis of datasets in which two or more
       tightly-linkedloci (for instance SNPs in the same gene) have been
       typed.<span style='mso-spacerun:yes'>  </span>ADMIXMAP allows for
       allelic association (if the distance between loci is coded as zero) and
       models the unobserved haplotypes.&nbsp; </span><o:p></o:p></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l2 level1 lfo1;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>STRUCTURE assumes that the markers have not
       been selected to be highly informative for ancestry, and the authors
       recommend that a model which allows for correlation between the allele
       frequencies in each subpopulation is specified. ADMIXMAP does not in
       general assume any correlation between the allele frequencies in each
       subpopulation.<span style='mso-spacerun:yes'>  </span></span></li>
  </ol>
  <p>&nbsp;</p>
  <p><a name="Running_the_program:"><span style='font-size:13.5pt;font-family:
  Arial;color:#214772'>Getting started</span></a></p>
  <p><span style='font-family:Arial'>ADMIXMAP runs as a console application.<span
  style='mso-spacerun:yes'>  </span>We are no longer supporting the Windows
  graphical front-end.<span style='mso-spacerun:yes'>  </span>Binaries
  (compiled executable files) are provided for Linux and Windows with i686
  processors.<span style='mso-spacerun:yes'>  </span>For other platforms, the
  source code and instructions for compiling are available from our CVS
  repository with the following settings</span> </p>
  <p><span style='font-family:Arial'>authentication: pserver<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>path: /dsk3/cvs<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>host address: bioinf.ucd.ie<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>CVS username: anonymous<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>password: &lt;blank&gt;<o:p></o:p></span></p>
  <p><span style='font-family:Arial'>If you have downloaded a binary, or
  compiled the program yourself, and placed the ADMIXMAP executable in a
  directory that lies on your search path, the program can be invoked directly by
  typing <o:p></o:p></span></p>
  <p><span style='font-family:Arial'><span style='mso-spacerun:yes'>     
  </span><i style='mso-bidi-font-style:normal'>admixmap</i> <i
  style='mso-bidi-font-style:normal'>&lt;--optionname=value&gt; &lt;--optionname=value&gt;
  <span style='mso-spacerun:yes'> </span>…</i><o:p></o:p></span></p>
  <p><span class=GramE><span style='font-family:Arial'>where</span></span><span
  style='font-family:Arial'> user options are specified on the command line.<span
  style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p><span style='font-family:Arial'>In practice, this is not recommended as
  the list of user options is too long to be typed manually.<span
  style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p><span style='font-family:Arial'>Alternatively, the program can be invoked by
  typing <o:p></o:p></span></p>
  <p><span style='font-family:Arial'><span style='mso-spacerun:yes'>     
  </span><i style='mso-bidi-font-style:normal'>admixmap &lt;optionsfile&gt;<o:p></o:p></i></span></p>
  <p><span class=GramE><span style='font-family:Arial'>where</span></span><span
  style='font-family:Arial'> <i style='mso-bidi-font-style:normal'>optionsfile </i>contains
  statements of the form <i style='mso-bidi-font-style:normal'>optionname=value</i>,
  one line per optionname. <i style='mso-bidi-font-style:normal'><span
  style='mso-spacerun:yes'> </span></i><o:p></o:p></span></p>
  <p><span style='font-family:Arial'>To generate summary statistics, tables and
  graphics from the output files generated by ADMIXMAP you will need to run the
  script <i style='mso-bidi-font-style:normal'>AdmixmapOutput.R</i>. <span
  style='mso-spacerun:yes'> </span>This requires the <a
  href="http://www.r-project.org/">R</a> statistical package to be installed,
  together with the <a href="http://www.public-health.uiowa.edu/boa"> BOA library</a> of functions for analysis of Markov chain </span><st1:City><st1:place><span
    style='font-family:Arial'>Monte Carlo</span></st1:place></st1:City><span
  style='font-family:Arial'>.
  <o:p></o:p></span></p>
  <p><span style='font-family:Arial'>The most convenient way to run the entire
  analysis is to use a Perl script. If you are working in Windows, you should
  install <a href="http://www.activestate.com/Products/ActivePerl/"> ActivePerl</a> if not already installed.<span
  style='mso-spacerun:yes'>  </span>The Perl <span class=GramE>script specifies
  the user options, run ADMIXMAP with these user options on the command line,
  and invoke</span> R in batch mode to run the script <i style='mso-bidi-font-style:
  normal'>AdmixmapOutput.R</i> to process the output files.<span
  style='mso-spacerun:yes'>  </span>See, for example<span class=GramE>, <span
  style='mso-spacerun:yes'> </span>the</span> Perl script <i style='mso-bidi-font-style:
  normal'>admixmap.pl</i> provided with the test files. User options can be
  changed by editing this file. This requires only minimal knowledge of Perl. <span
  style='mso-spacerun:yes'> </span>The Perl script is especially convenient
  when running a batch of different analyses, as each run may require only a
  few user options to be changed. <o:p></o:p></span></p>
  <p><span style='font-family:Arial'>The Perl script admixmap.pl should be
  invoked </span><font face="Arial">from a</font><span style='font-family:Arial'> folder that has a
  sub-folder named &quot;data&quot; containing the data files. Results will be
  output to another sub-folder named &quot;results&quot;. <span
  style='mso-spacerun:yes'> </span><span style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
  <p>&nbsp;</p>
  <p><a name="Options_specified_by_the_user:"><span style='font-size:13.5pt;
  font-family:Arial;color:#214772'>User options:</span></a></p>
  <p><span style='font-family:Arial'>The program requires a list of options to
  be specified by the user either as command-line arguments, or in a text file the
  name of which is given as a single argument to the <span class=GramE>program .&nbsp;</span>
  As explained above, the most convenient way to specify these arguments is to use
  the Perl script (&quot;admixmap.pl&quot;) provided.<span
  style='mso-spacerun:yes'>  </span>A list of these arguments is given in the
  following table.</span></p>
  <table class=MsoNormalTable border=1 cellspacing=0 cellpadding=0 width="853"
   style='mso-cellspacing: 0cm; mso-padding-alt: 7.15pt 7.15pt 7.15pt 7.15pt' height="7126">
   <tr style='mso-yfti-irow:0'>
    <td width="215" style='padding: 7.15pt' height="77" valign="top">
    <p><i><span style='font-family:Arial'>Option name, command-line argument.
    Required arguments in <b>bold</b></span></i><span style='font-family:Arial'>.</span></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="77">
    <p><i><span style='font-family:Arial'>Settings</span> </i></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:1'>
    <td style='padding: 7.15pt' width="215" height="74" valign="top">
    <p><b><i><span style='font-family:Arial'>samples</span></i></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="74">
    <p><span style='font-family:Arial'>Integer specifying total number of
    iterations of the Markov chain, <u>including burn-in</u>. A run of at least
    20 000 iterations is recommended for inference.</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:2'>
    <td style='padding: 7.15pt' width="215" height="92" valign="top">
    <p><b><i><span style='font-family:Arial'>burnin</span></i></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="92">
    <p><span style='font-family:Arial'>Integer specifying number of iterations
    for burn-in of the Markov chain, before posterior samples are output. A
    burn-in of at least 1000 iterations is recommended for inference.</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:3'>
    <td style='padding: 7.15pt' width="215" height="237" valign="top">
    <p><b><i><span style='font-family:Arial'>every</span></i></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="237">
    <p><span style='font-family:Arial'>Integer specifying the
    &quot;thinning&quot; of samples from the posterior distribution that are
    written to the output files, after the burn-in period.<span
    style='mso-spacerun:yes'>  </span>For example, if <i>every</i>=10, sampled
    values are written to the output files every 10 iterations<span
    class=GramE>&nbsp; We</span> recommend using a value of 5 to keep down the
    size of the output files.&nbsp; Sampling more frequently than this does not
    much improve the precision of results, because successive draws are not
    independent. <span style='mso-spacerun:yes'> </span>Thinning the output samples
    does not affect the calculation of ergodic averages or test statistics, which
    are based on all sampled values.<span style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:4'>
    <td style='padding: 7.15pt' width="215" height="39" valign="top">
    <i><span style="font-family: Arial">resultsdir</span></i>
    </td>
    <td style='padding: 7.15pt' width="592" height="39">
    <span style="font-family: Arial">Path of directory for output files. Default
    is 'results'.</span>
    </td>
   </tr>
   <tr style='mso-yfti-irow:4'>
    <td style='padding: 7.15pt' width="215" height="39" valign="top">
    <p><i><span style='font-family:Arial'>logfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="39">
    <p><span style='font-family:Arial'>Name of log file written by the
    program.</span> </p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:5'>
    <td width="829" colspan=2 style='padding: 7.15pt' height="77" valign="top">
    <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
    auto'><span style='font-family:Arial'>The program requires <b
    style='mso-bidi-font-weight:normal'>one</b> of the following four options,
    any one of which specifies the number of subpopulations in the model: <i>populations,
    allelefreqfile, priorallelefreqfile, </i>or <i>historicallelefreqfile</i>.&nbsp;
    These options are mutually exclusive</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:6'>
    <td style='padding: 7.15pt' width="215" height="184" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>populations</span></i><o:p></o:p></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="184">
    <p><span style='font-family:Arial'>Integer specifying number of
    subpopulations that have contributed to the admixed population under
    study.<span style='mso-spacerun:yes'>  </span>If specified as 1, the
    program fits a model based on a single homogeneous population.<span
    style='mso-spacerun:yes'>  </span>This option is not required if
    information about allele frequencies is supplied in <i>allelefreqfile</i>, <i>priorallelefreqfile</i>,
    or <i>historicallelefreqfile</i>, as the number of columns in any of these
    files defines the number of subpopulations in the model.<span
    style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:7'>
    <td style='padding: 7.15pt' width="215" height="255" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>allelefreqfile</span></i><o:p></o:p></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="255">
    <p><span style='font-family:Arial'>Pathname of file containing the allele
    frequencies of the genotyped loci for each subpopulation. When this option
    is specified, the model treats the allele frequencies as fixed constants. <o:p></o:p></span></p>
    <p><span style='font-family:Arial'>This option is obsolete, and retained
    only for backward compatibility.<span style='mso-spacerun:yes'> 
    </span>Instead, use option priorallelefreqfile to specify the allele
    frequencies, and specify option fixedalllelefreqs=1. <span
    style='mso-spacerun:yes'> </span>This allows you to use the same format for
    the allele frequency file, whether the allele frequencies are fixed, have a
    prior distribution with no dispersion, or are specified with a dispersion
    model.<span style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:8'>
    <td style='padding: 7.15pt' width="215" height="290" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>priorallelefreqfile</span></i><o:p></o:p></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="290">
    <p><span style='font-family:Arial'>Pathname of file containing parameters
    of the prior distributions for allele frequencies (or haplotype frequencies)
    at each compound locus in each subpopulation.<span
    style='mso-spacerun:yes'>  </span>Where allele frequencies have been
    estimated from a sample of unadmixed individuals, the prior distribution
    parameters for the corresponding subpopulation should be specified as the
    observed allele counts plus 0.5.&nbsp; Where no allele frequency data are
    available, specify the prior parameters as 0.5 for each allele
    (&quot;reference&quot; prior).<span style='mso-spacerun:yes'>  </span>When
    this option is specified, the program fits a model in which the allele
    frequencies in each subpopulation are estimated simultaneously from the
    unadmixed samples and the admixed sample under study</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:9'>
    <td style='padding: 7.15pt' width="215" height="164" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>historicallelefreqfile</span></i><o:p></o:p></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="164">
    <p><span style='font-family:Arial'>Pathname of file containing observed
    allele counts at the genotyped loci from samples of unadmixed individuals
    in each subpopulation.&nbsp; When this option is specified, the program
    fits a model that allows the &quot;historic&quot; allele frequencies in the
    unadmixed population to vary from the corresponding ancestry-specific
    allele frequencies in the admixed population under study</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:10'>
    <td style='padding: 7.15pt' width="807" height="5" colspan="2" valign="top">
    </td>
   </tr>
   <tr style='mso-yfti-irow:10'>
    <td style='padding: 7.15pt' width="215" height="58" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>locusfile,
    genotypesfile</span></i></b><i><span style='font-family:Arial'>,
    outcomevarfile, covariatesfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="58">
    <p><span style='font-family:Arial'>Pathnames of other input files: details
    of file formats below under <a href="#Input File Options"> Input files</a></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:11'>
    <td style='padding: 7.15pt' width="807" height="11" colspan="2" valign="top">
    </td>
   </tr>
   <tr style='mso-yfti-irow:11'>
    <td style='padding: 7.15pt' width="215" height="115" valign="top">
    <p><i><span style='font-family:Arial'>targetindicator</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="115">
    <p><font face="Arial">Integer specifying column in <i>outcomevarfile</i> that contains the
    outcome variable to be modelled. This column number should be specified as
    an offset from column 1: thus to select the variable in column 1, specify <i>targetindicator</i>=0.<span
    style='mso-spacerun:yes'>  </span>The default is 0.<span
    style='mso-spacerun:yes'>  </span></font></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:12'>
    <td style='padding: 7.15pt' width="215" height="1331" valign="top">
    <p><b style='mso-bidi-font-weight:normal'><i><span style='font-family:Arial'>analysistypeindicator</span></i><o:p></o:p></b></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="1331">
    <p><span style='font-family:Arial'>Integer specified as one of the
    following:-</span> </p>
    <p><span style='font-family:Arial'>0: Affected-only design</span></p>
    <p><span style='font-family:Arial'>1: No outcome variable</span></p>
    <p><span lang=FR style='font-family:Arial;mso-ansi-language:FR'>2: Model
    with continuous outcome variable (quantitative trait).&nbsp; </span><span
    style='font-family:Arial'>The program fits a linear regression model, with
    admixture as explanatory variable, unless option <i>admixturescorefile</i>
    is specified.<span style='mso-spacerun:yes'>  </span></span></p>
    <p><span style='font-family:Arial'>3: Model with binary outcome variable
    (e.g. unaffected / affected, coded<span class=GramE>&nbsp; as</span> 0 /
    1).<span style='mso-spacerun:yes'>  </span>The program fits a logistic
    regression model unless option <i>admixturescorefile</i> is
    specified.&nbsp; This option is appropriate for a cross-sectional study of
    a binary trait, a case-control study, or a cohort study in which
    individuals have been classified by disease status only once at follow-up.</span></p>
    <p><span style='font-family:Arial'>5: Model with two outcome variables,
    assumed to be independent given individual admixture proportions and
    covariates.<span style='mso-spacerun:yes'>  </span>With this option, the
    program can fit two different regression models, and the posterior
    distribution of individual admixture is conditioned on both outcome variables</span></p>
    <p class=MsoNormal><font face="Arial">-1: Model for a single individual whose parental admixture
    proportions and sum of intensities are estimated directly from the priors
    on these parameters, without making inference about the distribution of these
    parameters in the population (as in the standard hierarchical model). <span
    style='mso-spacerun:yes'> </span>With this option, the marginal likelihood
    of the model is calculated.</font> </p>
    <p class=MsoNormal><font face="Arial"><span style='mso-spacerun:yes'> </span>Estimates of
    individual admixture are not supplied.<span style='mso-spacerun:yes'> 
    </span>This model assumes no outcome variable.<span
    style='mso-spacerun:yes'>  </span>The priors on parental admixture
    proportions can be supplied with options <i style='mso-bidi-font-style:
    normal'>initalpha0</i> and <i style='mso-bidi-font-style:normal'>initalpha1</i>,
    and the prior on the sum of intensities parameter can be supplied with
    option <i style='mso-bidi-font-style:normal'>sumintensities. <o:p></o:p></i></font></p>
    <p class=MsoNormal><font face="Arial">The <span class=GramE>value <span
    style='mso-spacerun:yes'> </span><span style='mso-bidi-font-style:italic'>-</span></span><span
    style='mso-bidi-font-style:italic'>1 for this indicator can also be used
    with a dataset consisting of a single target individual and a sample of
    individuals from a population. The target individual's data should be on
    first row of genotypes file. The program will use data from the population
    sample to infer model parameters at population level, and calculate the
    marginal likelihood for the hypothesis that the target individual
    originates from the same population as the other individuals sampled. <o:p></o:p></span></font></p>
    <p><font face="Arial"><o:p>&nbsp;</font><span style='font-family:Arial'>-2: as for -1, but with estimates of
    individual admixture supplied in <i style='mso-bidi-font-style:normal'>mlefile</i>.
    This saves time when the only purpose of the analysis is to evaluate the
    marginal likelihood of the model.<span style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
    <p><span style='font-family:Arial'>-3: Model for a collection of
    individuals in which, as for -1 and -2, the admixture proportions of each
    individual’s parents, and the sum of intensities on each parental gamete,
    are statistically independent given the priors on these parameters.<span
    style='mso-spacerun:yes'>  </span><span style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
    <p><span style='font-family:Arial'>This option is useful in two situations:
    (1) when you already have strong prior information about the distribution
    of admixture in the population from which the individuals have been sampled,
    and want to specify a Dirichlet prior for each individual’s parental
    admixture proportions using the option initalpha0; or (2) when you want to
    calculate the marginal likelihood of the model given the genotype data on
    each individual.</span> </p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:13'>
    <td style='padding: 7.15pt' width="215" height="131" valign="top">
    <p class=MsoNormal><span style='font-family:Arial'>randommatingmodel</span>
    </p>
    </td>
    <td style='padding: 7.15pt' width="592" height="131">
    <p><span style='font-family:Arial'>0 - assortative mating model (admixture
    proportions the same in both parents)</span> </p>
    <p><span style='font-family:Arial'>1 - random mating model</span> </p>
    <p><span style='font-family:Arial'>The default is 0. </span><span
    style='mso-spacerun:yes'> </span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:14'>
    <td style='padding: 7.15pt' width="215" height="223" valign="top">
    <p class=MsoNormal><span style='font-family:Arial'>globalrho</span> </p>
    </td>
    <td style='padding: 7.15pt' width="592" height="223">
    <p><span style='font-family:Arial'>0 - the sum of intensities parameter </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'> is allowed to vary between <span class=GramE>individuals,</span>
    or between gametes if a random mating model is specified).<span
    style='mso-spacerun:yes'>  </span>This specifies a hierarchical model, with
    a gamma distribution for the variation of <span
    style='mso-spacerun:yes'> </span></span><span style='font-family:Symbol;
    mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
    Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
    style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'> between individuals specified as below. </span></p>
    <p><span style='font-family:Arial'>1 - the sum of intensities </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'> is modelled as a global parameter, set to be the
    same on all parental gametes</span> </p>
    <p><span style='font-family:Arial'>The default is 1</span> </p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:15'>
    <td style='padding: 7.15pt' width="215" height="551" valign="top">
    <p class=MsoNormal><span style='font-family:Arial'>sumintensities</span> </p>
    </td>
    <td style='padding: 7.15pt' width="592" height="551">
    <p><span style='font-family:Arial'>The sum of intensities parameter </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'> has a Gamma prior distribution, with shape
    parameter </span><span style='font-family:Symbol;mso-ascii-font-family:
    Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:Arial;mso-char-type:
    symbol;mso-symbol-font-family:Symbol'><span style='mso-char-type:symbol;
    mso-symbol-font-family:Symbol'>a</span></span><span style='font-family:
    Arial'> and scale <span class=GramE>parameter </span></span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>b</span></span><span
    style='font-family:Arial'>.</span> <span
    style='mso-spacerun:yes'> </span><font face="Arial">This option specifies the shape parameter
    of the prior distribution, with the scale parameter set to 1.<span
    style='mso-spacerun:yes'>  </span>Thus, for instance, to model an African-American
    population, for which we have prior information that the sum of intensities
    parameter is about 6 per morgan, we could specify</font> </p>
    <p><font face="Arial">sumintensities = 6</font> </p>
    <p><font face="Arial">This specifies the prior for the sum of intensities parameter</font> <span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'> as <span class=GramE>Gamma(</span>6, 1) which
    has mean 6 and variance 1. <span style='mso-spacerun:yes'> </span></span></p>
    <p><span style='font-family:Arial'>98 specifies </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>a</span></span><span
    style='font-family:Arial'> = </span><span style='font-family:Symbol;
    mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
    Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
    style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>b</span></span><span
    style='font-family:Arial'>= 0 (flat prior on log </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'>)</span> </p>
    <p><span style='font-family:Arial'>99 specifies </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>a</span></span><span
    style='font-family:Arial'> = 1, </span><span style='font-family:Symbol;
    mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
    Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
    style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>b</span></span><span
    style='font-family:Arial'> = 0 (flat prior on </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>r</span></span><span
    style='font-family:Arial'>)</span> </p>
    <p><font face="Arial">The default, if this option is not specified, is equivalent to
    sumintensities = 98.<span style='mso-spacerun:yes'>   </span>Where there is not enough data for
    reliable inference of the sum of intensities parameter, it is often useful to
    specify that the prior distribution should be truncated at some upper limit
    of plausible values, using the option <i style='mso-bidi-font-style:normal'>truncationpoint</i>.</font>
    </p>
    <p><font face="Arial">This option is not valid if <i style='mso-bidi-font-style:normal'>globalrho</i>=0
    has been specified unless the model has been specified without a
    hierarchical model for the sum of intensities, as with
    analysistypeindicator=-1, -2 or -3. <span style='mso-spacerun:yes'>  </span></font><o:p></o:p></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:16'>
    <td style='padding: 7.15pt' width="215" height="54" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>truncationpoint<o:p></o:p></span></i></p>
    </td>
    <td style='border-style: none; border-width: medium; padding: 7.15pt' width="594" height="54">
    <p><span style='font-family:Arial'>Upper value at which the prior for the sum
    of intensities parameter should be truncated<o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:17'>
    <td style='padding: 7.15pt' width="215" height="132" valign="top">
    <p class=MsoNormal><span style='font-family:Arial'>fixedallelefreqs</span> </p>
    </td>
    <td style='padding: 7.15pt' width="592" height="132">
    <p><span style='font-family:Arial'>1 - specifies that <i style='mso-bidi-font-style:
    normal'>priorallelefreqfile</i> contains fixed allele frequencies</span> </p>
    <p><span style='font-family:Arial'>0 - otherwise</span> </p>
    <p><font face="Arial">The default is 0</font></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:18'>
    <td style='padding: 7.15pt' width="215" height="149" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>coutindicator<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="149">
    <p><span style='font-family:Arial'>1 – <span class=GramE>sampled <span
    style='mso-spacerun:yes'> </span>values</span> of population parameters
    will be output to the console with frequency specified by <i
    style='mso-bidi-font-style:normal'>every.<o:p></o:p></i></span></p>
    <p><span style='font-family:Arial'>0<i style='mso-bidi-font-style:normal'> –
    </i>sampled values of population parameters will not be output to the
    console (but will be written to paramfile)<o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:19'>
    <td style='padding: 7.15pt' width="215" height="56" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>seed<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="56">
    <p><span class=GramE><span style='font-family:Arial'>can</span></span><span
    style='font-family:Arial'> be used to specify a seed for the random number
    generator. <o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:20'>
    <td style='padding: 7.15pt' width="215" height="705" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>etapriorfile<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="705">
    <p><span style='font-family:Arial'>File containing parameters of the gamma
    prior distribution specified for the allele frequency dispersion parameter </span><span
    style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
    Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
    Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>x</span></span><span
    style='font-family:Arial'> in each subpopulation.<span
    style='mso-spacerun:yes'>  </span>This option can be used only when a
    dispersion model has been specified with the option <i style='mso-bidi-font-style:
    normal'>historicallelefreqfile</i>. <span
    style='mso-spacerun:yes'> </span>This is useful when there are not enough
    data for the dispersion parameter to be inferred from the data, and we want
    to use prior information from population genetics. <span
    style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
    <p><span style='font-family:Arial'>This file has one row for each
    subpopulation (in the same order as the order of subpopulations by columns in
    historicallelefreqfile, and two columns specifying the shape and location
    parameters of the gamma distribution.<span style='mso-spacerun:yes'> 
    </span>Thus, for a sample from an African-American population, in which <i
    style='mso-bidi-font-style:normal'>historicallelefreqfile</i> contains
    counts of alleles in samples of modern west Africans (in the first column)
    and Europeans (in the second column), we might specify an etaprior file containing
    these two lines:- <o:p></o:p></span></p>
    <p><span style='font-family:Arial'>50<span style='mso-tab-count:1'>       </span>1<o:p></o:p></span></p>
    <p><span style='font-family:Arial'>500<span style='mso-tab-count:1'>     </span>1<o:p></o:p></span></p>
    <p><span style='font-family:Arial'>This specifies a prior with mean 50 for the
    parameter for dispersion of allele frequencies between modern unadmixed west
    Africans and the African gene pool in African-Americans, and a prior with
    mean 500 and variance 500 for the parameter for dispersion of allele
    frequencies between modern unadmixed Europeans and the European gene pool
    in African-Americans. <span style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
    <p><span style='font-family:Arial'>The dispersion parameter is related to the
    fixation index F<sub>ST</sub> by </span><span style='font-family:Symbol;
    mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
    Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
    style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>x</span></span><span
    style='font-family:Arial'> = (<span class=GramE>1 <span
    style='mso-spacerun:yes'> </span>+</span> F<sub>ST</sub>) / F<sub>ST</sub>,
    so values of 50 and 500 for </span><span style='font-family:Symbol;
    mso-ascii-font-family:Arial;mso-hansi-font-family:Arial;mso-bidi-font-family:
    Arial;mso-char-type:symbol;mso-symbol-font-family:Symbol'><span
    style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>x</span></span><span
    style='font-family:Arial'> correspond roughly to values of 0.02 and 0.002
    for F<sub>ST</sub>. <o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:21'>
    <td style='padding: 7.15pt' width="215" height="512" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>initalpha0,
    initalpha1<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="512">
    <p><span style='font-family:Arial'>These options are used only when the
    model has been specified without a hierarchical model of population
    admixture, as when <i style='mso-bidi-font-style:normal'>analysistypeindicator</i>=-1,
    -2, or -3.<span style='mso-spacerun:yes'>  </span>Each of these two options
    can be used to specify a Dirichlet parameter vector for parental admixture proportions.<span
    style='mso-spacerun:yes'>  </span>The parameter vector is specified as a
    string of numbers separated by commas.<span style='mso-spacerun:yes'> 
    </span>For instance, with a model based on 3 subpopulations:- <span
    style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
    <p><span style='font-family:Arial'>initalpha0 = “2, 8, 3.5” <o:p></o:p></span></p>
    <p><span class=GramE><span style='font-family:Arial'>would</span></span><span
    style='font-family:Arial'> specify the prior for parental admixture
    proportions (or the paternal gamete if option randommatingmodel=1 has been
    specified) with parameter vector c(2, 8, 3.5).<span
    style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
    <p><span class=GramE><span style='font-family:Arial'>initalpha1</span></span><span
    style='font-family:Arial'> can be used similarly to specify the prior for maternal
    admixture proportions if optionrandommatingmodel=1 has been specified. <span
    style='mso-spacerun:yes'> </span><span style='mso-spacerun:yes'> </span><o:p></o:p></span></p>
    <p class=MsoNormal style='margin-bottom:12.0pt'><span style='font-family:
    Arial;mso-bidi-font-style:italic'>For example, &quot;initalpha0 = 1<span
    class=GramE>,1,0</span>&quot; and &quot;initalpha1 = 1,1,1&quot; would
    specify that one parent has 2-way admixture (between subpopulations 1 and
    2) and the other has 3-way admixture between subpopulations .</span><span
    style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
    <p><span style='font-family:Arial'><o:p>&nbsp;</o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:22'>
    <td style='padding: 7.15pt' width="215" height="74" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>reportedancestry<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="74">
    <p><span style='font-family:Arial'>not fully tested or documented: allows prior
    information about each individual’s ancestry to be specified in the model<o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:23'>
    <td style='padding: 7.15pt' width="215" height="75" valign="top">
    <p><i><span style='font-family:Arial;mso-bidi-font-weight:bold'>xonlyanalysis<o:p></o:p></span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="75">
    <p><span style='font-family:Arial'>1 – specifies that all loci are on X
    chromosome<o:p></o:p></span></p>
    <p><span style='font-family:Arial'>0 (default value) - otherwise <o:p></o:p></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:24'>
    <td style='padding: 7.15pt' width="215" height="229" valign="top">
    <p><span class=GramE><i><span style='font-family:Arial'>paramfile</span></i></span><span
    style='font-family:Arial'><i>,</i></span> <span class=GramE><i><span style="font-family: Arial">regparamfile</span></i></span><span class=GramE><i><span
    style='font-family:Arial'>, dispparamfile,</span></i></span> <i><span
    style='font-family:Arial'>ergodicaveragefile, indadmixturefile,
    allelefreqoutputfile, admixturescorefile, allelicassociationscorefile,
    ancestryassociationscorefile, affectedsonlyscorefile, allelefreqscorefile,
    stratificationtestfile, dispersiontestfile, haplotypeassociationscorefile.</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="229">
    <p><span style='font-family:Arial'>Pathnames of output files, details of
    file formats in <a href="#Output_files:">Output files</a>.</span> </p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:25'>
    <td style='padding: 7.15pt' width="215" height="348" valign="top">
    <p><i><span style='font-family:Arial'>admixturescorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="348">
    <p><span style='font-family:Arial'>Pathname of file to which results of a
    score test for the association of the trait with individual admixture will
    be written. This option is valid only if an outcome variable
    has been specified.<span style="mso-spacerun: yes"> </span>This option is
    used only to obtain a formal test of the null hypothesis of no association
    between the trait and individual admixture.&nbsp; If <i>admixturescorefile </i>is
    specified, the regression model will not include individual admixture
    proportions as explanatory variables, and tests for allelic association or
    linkage will not be adjusted for the effect of individual admixture.&nbsp; </span></p>
    <p><span style='font-family:Arial'>Unless option <i>admixturescorefile </i>is
    specified or <i>analysistypeindicator</i> is specified as 0 or 1, the
    program will fit a regression model with the outcome variable as dependent
    variable and individual admixture proportions (plus any covariates
    specified in<span style='mso-spacerun:yes'>  </span>inputfile) as
    explanatory variables.&nbsp; </span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:26'>
    <td width="829" colspan=2 style='padding: 7.15pt' height="38" valign="top">
    <p><span style='font-family:Arial'>The options below specify additional
    tests or output,but do not change the model itself</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:27'>
    <td style='padding: 7.15pt' width="215" height="74" valign="top">
    <p><i><span style='font-family:Arial'>allelicassociationscorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="74">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing score
    tests for association of the outcome variable with alleles at each simple
    locus, adjusting for individual admixture.</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:28'>
    <td style='padding: 7.15pt' width="215" height="92" valign="top">
    <p><i><span style='font-family:Arial'>haplotypeassociationscorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="92">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing score
    tests for association of the outcome variable with haplotypes for all
    compound loci containing haplotypes, adjusting for individual admixture.</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:29'>
    <td style='padding: 7.15pt' width="215" height="291" valign="top">
    <p><i><span style='font-family:Arial'>ancestryassociationscorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="291">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing score
    tests at each compound locus for linkage with genes underlying ethnic
    variation in the trait. This is a test for association of the trait with
    locus ancestry, adjusting for individual admixture and covariates. This
    test should be used in a cross-sectional or cohort study design. For a case-control
    study of a rare disease, the affected-only test below has greater
    statistical power. <o:p></o:p></span></p>
    <p><span style='font-family:Arial'>Calculation of these tests requires a
    long sampling run.<span style='mso-spacerun:yes'>  </span>This will be
    fixed in a later version by implementing a more efficient algorithm for calculating
    the test statistic, similar to that used for the affected-only test. </span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:30'>
    <td style='padding: 7.15pt' width="215" height="201" valign="top">
    <p><i><span style='font-family:Arial'>affectedonlyscorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="201">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing score
    tests at each compound locus for linkage with ancestry, based on comparing
    the observed and expected proportions of gene copies at this locus that
    have ancestry from each subpopulation.<span style='mso-spacerun:yes'> 
    </span>This test is calculated from affected individuals only: individuals
    are their own controls.&nbsp; Even when the sample includes both cases and
    controls, this test is more powerful than the regression model score test
    in <i>ancestryassociationscorefile</i> if the disease is rare<i>.&nbsp; </i></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:31'>
    <td style='padding: 7.15pt' width="215" height="93" valign="top">
    <p><i><span style='font-family:Arial'>allelefreqscorefile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="93">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing score
    tests of mis-specified<span class=GramE>&nbsp; ancestry</span> specific
    allele frequencies.&nbsp; This option is valid only when the allele
    frequencies are fixed, i.e. when option <i>allelefreqfile</i> is specified.</span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:32'>
    <td style='padding: 7.15pt' width="215" height="130" valign="top">
    <p><i><span style='font-family:Arial'>allelefreqoutputfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="130">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing
    samples from the posterior distribution of ancestry-specific allele
    frequencies. Valid only when the allele frequencies are specified as random
    variables, i.e. when one of the two options <i>priorallelefreqfile</i> or <i>historicallelefreqfile</i>
    is specified.<span style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:33'>
    <td style='padding: 7.15pt' width="215" height="74" valign="top">
    <p><i><span style='font-family:Arial'>stratificationtestfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="74">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing test
    for residual population stratification (stratification not accounted for by
    the fitted model).&nbsp; </span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:34'>
    <td style='padding: 7.15pt' width="215" height="183" valign="top">
    <p><i><span style='font-family:Arial'>dispersiontestfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="183">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing test
    for dispersion of allele frequencies between the unadmixed populations
    sampled and the corresponding ancestry-specific allele frequencies in the
    admixed population under study.<span style='mso-spacerun:yes'>  </span>This
    is evaluated for each subpopulation at each locus, and as a global test
    over all loci.&nbsp; This option is valid only if option <i>priorallelefreqfile</i>
    is specified. The results are &quot;Bayesian p-values&quot;, as above.<span
    style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
   <tr style='mso-yfti-irow:35;mso-yfti-lastrow:yes'>
    <td style='padding: 7.15pt' width="215" height="230" valign="top">
    <p><i><span style='font-family:Arial'>FSToutputfile</span></i></p>
    </td>
    <td style='padding: 7.15pt' width="592" height="230">
    <p><span style="font-family: Arial">N</span><span style='font-family:Arial'>ame of output file containing
    posterior samples of Wright's F<sub>ST</sub> for each locus and each
    subpopulation.<span style='mso-spacerun:yes'>  </span>This option is valid
    only if option <i>historicallelefreqfile</i> is specified. For each
    subpopulation and each locus, the F<sub>ST</sub> statistic measures the
    variation between the allele frequencies in the &quot;historic&quot;
    population (the unadmixed population that was sampled to obtain the allele
    counts given in the file <i>historicallelefreqfile)</i> and the
    corresponding ancestry-specific allele frequencies in the admixed
    population under study.<span style='mso-spacerun:yes'>  </span></span></p>
    </td>
   </tr>
  </table>
  <p style="line-height: 100%; margin-top: 0; margin-bottom: 0">&nbsp;</p>
  <p style="line-height: 100%; margin-top: 0; margin-bottom: 0"><span style="font-size: 13.5pt; font-family: Arial; color: #214772"><a name="Input File Options">Input
  File Options</a></span><span style="font-family: Arial"><br>
  </span></p>
  <p style="line-height: 100%; margin-top: 0; margin-bottom: 0"><span class=GramE><i><span style="font-family: Arial" class="GramE">genotypesfile</span></i></span><br>
  <span style="font-family: Arial">The first row of the file is a header row
  listing locus names, enclosed in quotes and separated by spaces. Locus names
  should be exactly the same as in the file locusfile. Loci must be ordered by
  their map positions on the genome. Each subsequent row contains genotype data
  for a single individual. Each line contains the individual ID, followed by
  observed genotypes at each locus. Genotypes are coded as strings enclosed in
  quotes, separated by a single space.&nbsp; Each string consists of two numbers
  separated by a comma, with no spaces.&nbsp;&nbsp; Where there are a alleles at
  a locus, the alleles should be coded as numbers from 1 to a.&nbsp; Missing
  genotypes are coded as empty strings (&quot;&quot;).&nbsp;<br>
  For compatibility with existing datasets, we plan to change this file format
  to one more similar to the PEDFILE format used with LINKAGE.&nbsp;<br>
  </span></p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">locusfile</span></i></span><br>
  <span style="font-family: Arial">File contains information about each simple
  locus: that is, each locus that is typed.&nbsp; The first row of the file is
  ignored by the program, and can be used as a header. Each subsequent row
  contains values of three variables: locus name, number of alleles at this
  locus, and genetic map distance in Morgans between this locus and the previous
  locus.&nbsp; Loci must be ordered by their map positions on the genome. Locus
  names should begin with a letter, and contain only alphanumeric characters (no
  spaces, dots or hyphens).&nbsp; If the previous locus is unlinked, the genetic
  map distance in morgans should be coded as 100. For two or more loci that are
  so close together that they should be analysed as a single compound locus (as
  with DRD2Bcl and DRD2Taqd) in the example below, map distance should be coded
  as 0.<br>
  The website <a href="http://actin.ucd.ie/cgi-bin/rs2cm.cgi">http://actin.ucd.ie/cgi-bin/rs2cm.cgi</a>
  can be used to obtain the genetic map positions (in cM)<br>
  of a list of SNPs, which are then specified (as distance from last, in morgans) in the locusfile.</span></p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">allelefreqfile</span></i></span><br>
  <span style="font-family: Arial">This file contains the ancestry-specific
  allele frequencies at each compound locus in each ancestral subpopulation. The
  first row contains headers in quotes, separated by spaces. The first string in
  this row is ignored. Subsequent strings in the first row specify the names of
  the ancestral subpopulations contributing to the admixed population under
  study.&nbsp; Subsequent rows specify the ancestry-specific allele frequencies
  (usually estimated by from sampling modern descendants of the subpopulations
  that underwent admixture. The first column in each row gives the name of the
  compound locus, in quotes<br>
  For biallelic loci, only the frequency of allele 1 in each population is
  specified. For each locus with a alleles, there are a - 1 rows specifying
  frequencies of alleles 1 to a - 1.<br>
  Where two or more loci are to be analysed as a single haplotype, the
  ancestry-specific frequency of each haplotype must be specified. Thus in the
  example files below, there are two SNPs in the DRD2 gene, giving four possible
  haplotypes) and four lines specifying the ancestry-specific frequencies of
  haplotypes 11, 12, 21, 22. The loci in the haplotype are ordered by their map
  position on the genome, and the haplotypes are ordered by incrementing a
  counter from right to left. For instance if there were three loci A, B, C,
  with 4, 2 and 3 alleles respectively, the haplotypes would be listed in the
  following order: 111, 112, 113, 121, 122, 123, 211, ..., 422, 423.<br>
  Note: Use of this file is not recommended and is supported only for backward
  compatibility with previous versions. &nbsp;Instead you can specify the allele
  frequencies in priorallelefreqfile or historicallelefreqfile as fixed, with
  option fixedallelefreqs = 1</span><br>
  </p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">priorallelefreqfile</span></i></span><br>
  <span style="font-family: Arial">This file contains parameter values for the
  Dirichlet prior distribution of the allele or haplotype frequencies at each
  compound locus in each subpopulation.&nbsp; At each compound locus with k
  alleles or k possible haplotypes, a Dirichlet prior distribution is specified
  by a vector of k positive numbers.&nbsp;&nbsp; Where these alleles or
  haplotypes have been counted directly in samples from an unadmixed
  subpopulation, the parameter values should be specified as 0.5 plus the
  observed counts of each allele.&nbsp;&nbsp; Where no information is available
  about allele or haplotype frequencies at a compound locus in a subpopulation,
  or no copies of the allele have been observed in the sample from that
  subpopulation, specify 0.5 in the corresponding cells.&nbsp;&nbsp; Specifying
  0.5 in all cells, with columns for b subpopulations, is equivalent to
  specifying the option populations = b<br>
  Where haplotype frequencies at a compound locus have been estimated from
  unordered genotypes, the user should supply the parameters of the Dirichlet
  distribution that most closely approximates the posterior distribution of
  haplotype frequencies given the observed genotypes and a reference prior, as
  described above.&nbsp; The first row is a header row, consisting of strings in
  quotes, separated by spaces. The first string in this row is ignored, and the
  subsequent strings specify the names of the ancestral subpopulations
  contributing to the admixed population).<br>
  After the header row, there is one row for each allele (or haplotype) at each
  compound locus.&nbsp; The first column in each row gives the name of the
  compound locus in quotes. &nbsp;Subsequent columns give the prior parameters
  for the frequency of the allele (or haplotype) in each subpopulation,
  separated by a single space.&nbsp;<br>
  If the compound locus consists of two or more simple loci, (see notes above),
  the rows list prior parameters for the haplotypes in the order defined by
  incrementing a counter from right to left. For instance if there were three
  loci A, B, C, with 4, 2 and 3 alleles respectively, the haplotypes would be
  listed in the following order: 1-1-1, 1-1-2, 1-1-3, 1-2-1, 1-2-2, 1-2-3,
  2-1-1, ..., 4-2-2, 4-2-3.&nbsp; Estimated counts should be given for all
  possible haplotypes, however rare.&nbsp; The program will include all possible
  haplotypes in the model, but will omit rare haplotypes when constructing test
  statistics.&nbsp;<br>
  </span></p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">historicallelefreqfile</span></i></span><br>
  <span style="font-family: Arial">This file contains observed counts of alleles
  or haplotypes at each compound locus in samples from unadmixed
  subpopulations.&nbsp; The format of this file is exactly the same as the
  format of priorallelefreqfile described above.&nbsp; The only difference
  between the two files is that in historicallelefreqfile 0.5 is not added to
  the observed counts.&nbsp;&nbsp;<br>
  </span></p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">outcomevarfile</span></i></span><br>
  <span style="font-family: Arial">This file is required if
  analysistypeindicator has been specified as 2, 3 or 5.&nbsp;&nbsp; The file
  contains values of one or more outcome variables.&nbsp; After the header row,
  the file has one row per individual.&nbsp; Binary variables should be coded as
  1 = affected, 0 = unaffected. The header row contains the variable labels in
  quotes separated by spaces. A target file is required if analysistypeindicator
  has been specified as 2 (cross-sectional study with continuous outcome) or 3
  (binary outcome).&nbsp; If the file contains more than one outcome variable,
  the column containing the variable of interest should be specified by the
  command-line option targetindicator. Missing values are coded as #.<br>
  </span></p>
  <p><span class=GramE><i><span style="font-family: Arial" class="GramE">covariatesfile</span></i></span><br>
  <span style="font-family: Arial">This file contains values of covariates to be
  included in the regression model: it is used only if analysistypeindicator has
  been specified as 2, 3 or 5, and is optional even then.&nbsp; The header row
  contains covariate names in quotes, separated by spaces. Subsequent rows
  contain the observed values of these variables.&nbsp; For computational
  reasons, the values of the covariates should be centred about their sample
  means. Missing values are coded as #.</span></p>
  <p>&nbsp;</p>
  <p>&nbsp;</p>
  <p><a name="Output_files:"><span style='font-size:13.5pt;font-family:Arial;
  color:#214772'>Output files:</span></a></p>
  <p class=MsoNormal><span style='font-family:Arial'>Output files are formatted
  as either tab-delimited <span class=GramE>tables</span> with a header line (for
  2-way arrays) or as R objects (for 3- or 4-way arrays). Output files are
  written to the directory specified by <i>resultsdir.</i></span> </p>
  <p class=stylebefore5ptafter5pt><span class=GramE><i><span style='font-family:
  Arial'>paramfile</span></i></span><span style='font-family:Arial'> - Posterior draws
  of the following at intervals determined by option every: -</span></p>
  <ol start=1 type=1>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo2;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>Parameters of the Dirichlet distribution for
       parental admixture: one for each subpopulation</span></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l1 level1 lfo2;tab-stops:list 36.0pt'><span
       style='font-family:Arial'>Sum of intensities for the stochastic process
       of transitions of ancestry on hybrid chromosomes</span></li>
  </ol>
  <p><span class=GramE><i><span style='font-family:
  Arial'>regparamfile</span></i></span><span style='font-family:Arial'> - </span><span style="font-family: Arial">Posterior draws
  of intercept, slop and precision (the inverse of the residual variance) parameters in the regression model,
  at intervals determined by option every.</span></p>
  <p><span class=GramE><i><span style="font-family: Arial">disp</span></i></span><span class=GramE><i><span style='font-family:
  Arial'>paramfile</span></i></span><span style='font-family:Arial'> - </span><span style="font-family: Arial">Posterior
  draws of allele frequency dispersion parameters, one for each subpopulation, at intervals determined by option every.</span>
  <span
       style='font-family:Arial'>These are written only if option
       historicallelefreqfile has been specified.</span></p>
  <p class=MsoNormal><span style='font-family:Arial'>Median and 95% credible
  intervals for these parameters are written to the file
  PosteriorQuantiles.txt.</span> </p>
  <p><span class=GramE><i><span style='font-family:Arial'>indadmixturefile</span></i></span><i><span
  style='font-family:Arial'> </span></i><span style='font-family:Arial;
  mso-bidi-font-style:italic'>- Posterior draws of individual/gamete level
  variables, at intervals determined by option every written as an R object.
  The outputs to this file are, in the following order;</span><span
  style='mso-bidi-font-style:italic'> <o:p></o:p></span></p>
  <ol start=1 type=1>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l0 level1 lfo3;tab-stops:list 36.0pt'><span class=GramE><span
       style='font-family:Arial;mso-bidi-font-style:italic'>gamete</span></span><span
       style='font-family:Arial;mso-bidi-font-style:italic'> admixture
       proportions, ordered by subpopulations and then by gamete if a random
       mating model is specified. If an assortative mating model is specified
       only individual admixture proportions will be output.</span><span
       style='mso-bidi-font-style:italic'> <o:p></o:p></span></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l0 level1 lfo3;tab-stops:list 36.0pt'><span
       style='font-family:Arial;mso-bidi-font-style:italic'>gamete/individual
       sum-of-intensities if globalrhoindicator is false</span><span
       style='mso-bidi-font-style:italic'> <o:p></o:p></span></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l0 level1 lfo3;tab-stops:list 36.0pt'><span
       style='font-family:Arial;mso-bidi-font-style:italic'>predicted value of
       the outcome variable in the regression model</span><span
       style='mso-bidi-font-style:italic'> <o:p></o:p></span></li>
   <li class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:
       auto;mso-list:l0 level1 lfo3;tab-stops:list 36.0pt'><span class=GramE><span
       style='font-family:Arial;mso-bidi-font-style:italic'>paternal</span></span><span
       style='font-family:Arial;mso-bidi-font-style:italic'> and maternal
       haplotypes at this locus.</span><span style='mso-bidi-font-style:italic'>
       <o:p></o:p></span></li>
  </ol>
  <p class=MsoNormal><span style='font-family:Arial;mso-bidi-font-style:italic'>These
  values are written out for every individual <span class=GramE>at every
  iteration</span> This file is formatted to be read into R as a three-way
  array (indexed by variables, individuals, draws)<i>.</i></span><i> <o:p></o:p></i></p>
  <p class=stylebefore5ptafter5pt><span class=GramE><i><span style='font-family:
  Arial'>allelefreqoutputfile</span></i></span><i><span style='font-family:
  Arial'> - </span></i><span style='font-family:Arial;mso-bidi-font-style:italic'>Posterior
  draws of the ancestry-specific allele or haplotype frequencies for each state
  of ancestry at each compound locus, at intervals determined by option
  every.<span style='mso-spacerun:yes'>  </span>These results can be used to
  compute new parameters for the prior distributions specified in
  priorallelefreqfile which can be used in subsequent studies with independent
  samples</span><i><o:p></o:p></i></p>
  <p class=stylebefore5ptafter5pt><span class=GramE><i><span style='font-family:
  Arial'>ergodicaveragefile</span></i></span><i><span style='font-family:Arial'> - </span></i><span style='font-family:Arial;mso-bidi-font-style:italic'>Cumulative
  posterior means over all iterations (&quot;ergodic averages&quot;) for the
  variables in paramfile, regparamfile and dispparamfile, output at intervals of 10 </span><span
  style='font-family:Symbol;mso-ascii-font-family:Arial;mso-hansi-font-family:
  Arial;mso-bidi-font-family:Arial;mso-char-type:symbol;mso-symbol-font-family:
  Symbol;mso-bidi-font-style:italic'><span style='mso-char-type:symbol;
  mso-symbol-font-family:Symbol'>´</span></span><span style='font-family:Arial;
  mso-bidi-font-style:italic'>every iterations.&nbsp; Monitoring these ergodic <span
  class=GramE>averages allows</span> the user to determine whether the sampler
  has been run long enough for the posterior means to have been estimated
  accurately.<span style='mso-spacerun:yes'>  </span></span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>The output
  files<i> admixturescorefile, allelicassociationscorefile,
  ancestryassociationscorefile, affectedsonlyscorefile </i>contain results of
  score tests obtained by averaging over the posterior distribution.<span
  style='mso-spacerun:yes'>  </span>Each table of score test results, based on
  cumulative averages for the score and information over all posterior samples
  obtained after the burn-in period, is output at intervals of 10 × every.<span
  style='mso-spacerun:yes'>  </span>Monitoring these repeated updates allows
  the user to determine when the sampler has been run long enough for the test
  results to be computed accurately.<span style='mso-spacerun:yes'>  </span>For
  inference, only the last table in the output file, which is based on the entire
  posterior sample, is used.<span style='mso-spacerun:yes'>  </span>&nbsp;<span
  class=GramE>all</span> these files are formatted to be read into R as a
  three-way array (indexed by loci, test statistics, output number.<i><span
  style='mso-spacerun:yes'>  </span></i></span><i><o:p></o:p></i></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>For univariate
  null hypotheses (testing the effect of one allele, one haplotype, or one
  subpopulation against all others) the test statistic is the score divided by
  the square root of the observed information, which has a standard normal
  distribution under the null hypothesis.&nbsp;&nbsp; The percent of
  information extracted (the ratio of observed information to complete
  information) measures the information obtained about the parameter under
  test, in comparison the information that would be obtained if individual
  admixture, haplotypes at each locus, and gamete ancestry at each locus were
  measured without error.<span style='mso-spacerun:yes'>  </span><o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>For the
  affected-only score test, the missing information can be partitioned into two
  components: missing information about locus ancestry, and missing information
  about model <span class=GramE>parameters(</span>parental admixture) .<span
  style='mso-spacerun:yes'>  </span>These components are tabulated
  separately.<span style='mso-spacerun:yes'>  </span></span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>For composite
  null hypotheses, the score <b>U</b> is a vector, the observed information <b>V</b>
  is a matrix, and the test statistic (<b>UV<sup>-1</sup>U<sup>/</sup></b>) has
  a chi-squared distribution under the null hypothesis.&nbsp; </span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  class=GramE><i><span style='font-family:Arial'>admixturescorefile</span></i></span><span
  style='font-family:Arial;mso-bidi-font-style:italic'> - test for association
  of trait with individual admixture.<span style='mso-spacerun:yes'> 
  </span>The null hypothesis is no effect of individual admixture in a
  regression model, with covariates as explanatory variables if specified.<span
  style='mso-spacerun:yes'>   </span>The test statistic is computed for the
  effect of each subpopulation separately, with a summary chi-square test over
  all subpopulations if there are more than two subpopulations.&nbsp; </span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  class=GramE><i><span style='font-family:Arial'>allelicassociationscorefile</span></i></span><span
  style='font-family:Arial;mso-bidi-font-style:italic'> - tests for allelic
  association at each locus.&nbsp; The null hypothesis is no effect of the
  alleles or haplotypes in a regression analysis with individual admixture (and
  covariates if specified) as explanatory variables. The test statistic is
  computed for each allele or haplotype separately, with a summary chi-square
  statistic over all alleles or haplotypes at each locus if there are more than
  two alleles or haplotypes.&nbsp; Rare alleles or haplotypes are grouped
  together.<span style='mso-spacerun:yes'>   </span><br>
  This test is appropriate when testing for association of the trait with
  alleles or haplotypes in a candidate gene.<span style='mso-spacerun:yes'> 
  </span></span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=stylebefore5ptafter5pt><span class=GramE><i><span style='font-family:
  Arial'>ancestryassociationscorefile</span></i></span><span style='font-family:
  Arial;mso-bidi-font-style:italic'> - tests for linkage of each locus with
  genes underlying ethnic variation in disease risk or trait values.&nbsp; This
  is a test for association of the trait with ancestry at each compound locus,
  conditional on parental admixture.<span style='mso-spacerun:yes'>  </span>The
  null hypothesis is no effect of locus ancestry in a regression analysis with
  individual admixture (and covariates if specified) as explanatory
  variables.&nbsp; The test statistic is computed for the effect of each
  subpopulation separately, with a summary chi-square statistic over all
  subpopulations at each locus if there are more than two subpopulations.&nbsp;
  The proportion of information extracted depends upon the information content
  for ancestry of the marker locus and other nearby loci.<span
  style='mso-spacerun:yes'>  </span>This test is appropriate when the objective
  of the study is to exploit admixture to localize genes underlying ethnic
  variation in the trait value, using ancestry-informative markers rather than
  candidate gene polymorphisms.&nbsp; </span><span style='mso-bidi-font-style:
  italic'><o:p></o:p></span></p>
  <p class=stylebefore5ptafter5pt><span class=GramE><span class=grame><i><span
  style='font-family:Arial'>affectedsonlyscorefile</span></i></span></span><span
  style='font-family:Arial;mso-bidi-font-style:italic'> - tests for linkage of
  each locus with genes underlying the ethnic difference in disease risk, using
  only the affected individuals.&nbsp;&nbsp;&nbsp; The null hypothesis is that
  the risk ratio between populations that the locus accounts for is 1.&nbsp;
  This test statistic is computed for the effect of each subpopulation at each
  locus.&nbsp; The test compares at each locus the observed and expected
  proportion of gene copies that have ancestry from the high-risk
  subpopulation. This is the only test that can be used if the sample consists
  only of affected individuals. &nbsp;Even if a control group has been typed,
  for a rare disease the affected-only test is more efficient than the test
  given in ancestryassociationscorefile based on a regression model.<span
  style='mso-spacerun:yes'>  </span>This is because for a rare disease, the
  observed and expected proportion of gene copies that have ancestry from the high-risk
  subpopulation will not differ by very much in unaffected individuals.&nbsp; </span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  class=GramE><span class=grame><i><span style='font-family:Arial'>allelefreqscorefile</span></i></span></span><span
  style='font-family:Arial;mso-bidi-font-style:italic'> - tests for
  mis-specification of ancestry-specific allele frequencies. <br>
  This test is computed only if allele frequencies have been specified as fixed
  with option allelefreqfile.<span style='mso-spacerun:yes'>  </span>For each
  compound locus and each subpopulation, a score test is computed for the null
  hypothesis that the frequencies of all alleles have been specified correctly.<span
  style='mso-spacerun:yes'>  </span>A summary test over all k subpopulations is
  also computed at each locus.</span><span style='mso-bidi-font-style:italic'> &nbsp;</span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'>&nbsp;</p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span style="font-family: Arial" class="grame"><i>args.txt
  - </i>a list of the options used by the program. This is used by the R script
  to identify&nbsp;</span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span style="font-family: Arial" class="grame">output
  files and other information. This is writen to the location of the admixmap
  executable</span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span style="font-family: Arial" class="grame">and
  to <i>resultsdir</i>.</span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=stylebefore5ptafter5pt><span style='font-family:Arial;mso-bidi-font-style:
  italic'>An R script (AdmixmapOutput.R) is supplied that processes these
  output files to <span class=GramE><span class=grame>produce<span
  style='mso-spacerun:yes'>  </span>tables</span></span> of posterior
  quantiles, frequency plots of the posterior distribution, and plots of the
  cumulative posterior means for the variables that are output to paramfile.<span
  style='mso-spacerun:yes'>  </span>The R script also calculates a summary
  slope parameter for the effect of admixture from each subpopulation, versus
  the others.&nbsp;&nbsp; This R script is run automatically from the Perl
  script (admixmap.pl) that is supplied as a wrapper for the program.<span
  style='mso-spacerun:yes'> &nbsp; </span></span></p>
  <p class=stylebefore5ptafter5pt><span style='mso-bidi-font-style:
  italic'><o:p></o:p></span></p>
  <p class=stylebefore5ptafter5pt>&nbsp;</p>
  <p><font face="Arial"><a name="Interpretation of output from the program"><font color="#214772"><b>Interpretation of output from the program</b></font></a></font><o:p></o:p></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>These notes are based on
  the output produced by using the Perl script admixmap.pl to run the main
  program.<span style='mso-spacerun:yes'>  </span>Output files produced by the
  main are processed by the R script AdmixmapOutput.R.<span
  style='mso-spacerun:yes'> </span>This produces several text files, and a
  file plots.ps containing graphs in postscript format</span><span
  style='mso-bidi-font-style:italic'> <o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='mso-bidi-font-style:italic'>&nbsp;<o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'><u>Evaluating
  the sampler</u></span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>The adequacy of the
  burn-in period can be evaluated by the Geweke diagnostics in the boa
  output.<span style='mso-spacerun:yes'>   </span>If the burn-in period is
  adequate, the numbers in this table should have approximately a standard
  normal distribution.<span style='mso-spacerun:yes'>  </span></span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>The mixing of the MCMC
  sampler can be evaluated by examining the autocorrelation plots.&nbsp;
  Autocorrelation extending beyond 20 iterations (2 thinned draws if every = <span
  class=GramE>10 )</span> indicates slow mixing.<span
  style='mso-spacerun:yes'>  </span></span><span style='mso-bidi-font-style:
  italic'><o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>The adequacy of
  the total number of iterations can be evaluated by examining a plot of the
  statistic of interest calculated from all iterations since the end of the
  burn-in period, against the iteration number.<span
  style='mso-spacerun:yes'>   </span>Where inference is based on the mean of a
  parameter, this statistic is an &quot;ergodic average&quot; (cumulative
  average) over all iterations to that point.<span style='mso-spacerun:yes'> 
  </span>Plots of ergodic averages of the population-level parameters are given
  in file Plots.ps.&nbsp; </span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p><u><span style='mso-bidi-font-style:italic'>&nbsp;</span><span
  style='font-family:Arial;mso-bidi-font-style:italic'>Evaluating the fit of
  the model</span></u><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>The file <i>stratificationtestfile</i>
  contains results of a diagnostic test for residual population stratification
  that is not explained by the fitted model.<span style='mso-spacerun:yes'> </span>For details of how this test is calculated, and a discussion of how to
  interpret it, see Hoggart (2003).<span style='mso-spacerun:yes'> </span>The
  test is based on testing for allelic association between unlinked loci that
  is not explained by the model.<span style='mso-spacerun:yes'> </span>The <span
  class=GramE>results is</span> a &quot;Bayesian p-value&quot;:<span
  style='mso-spacerun:yes'>  </span>p &lt; 0.5 indicates lack of fit.<span style="mso-spacerun: yes">
  </span>The &quot;Bayesian p-value&quot; calculated
  by this test is more conservative than a classical p-value.<span
  style='mso-spacerun:yes'>&nbsp; </span>Our experience has been that a test
  p-value of 0.3 or less is fairly strong evidence for residual
  stratification.<span style='mso-spacerun:yes'>  </span>Where this statistic
  yields evidence of lack of fit, the model should be specified with more
  subpopulations, unless there is some other reason for lack of fit such as
  mis-specified allele frequencies.&nbsp; </span><span style='mso-bidi-font-style:
  italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>The file <i>dispersiontestfile</i>
  contains results of a diagnostic test for variation between the allele
  frequencies in the unadmixed populations that have been sampled to calculate
  the prior parameter values in priorallelefreqfile and the corresponding
  ancestry-specific allele frequencies in the admixed population under
  study.<span style='mso-spacerun:yes'>  </span>Again the results are
  &quot;Bayesian p-values&quot;, for which the deviation of the test p-value
  from its expected value of 0.5 does not provide an absolute measure of the
  strength of evidence for lack of fit.<span style='mso-spacerun:yes'> </span>For each subpopulation, the test statistic is calculated as a summary
  test over all loci and for each locus separately.<span
  style='mso-spacerun:yes'>  </span>Examination of the test statistic for each
  locus may reveal errors in coding, or errors in specifying the prior allele
  frequencies.&nbsp; </span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>The option <i>dispersiontestfile</i>
  is meaningful only where option priorallelefreqfile has been specified.<span
  style='mso-spacerun:yes'>  </span>Where allele frequencies have been
  specified as fixed, option <i>allelefreqscorefile</i> should be specified and
  the output file should be examined.<span style='mso-spacerun:yes'>  </span></span><span
  style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><span
  style='font-family:Arial;mso-bidi-font-style:italic'>No diagnostic test for
  lack of fit of the distribution of individual admixture proportions to the
  model is yet implemented.<span style="mso-spacerun: yes"> </span>However the
  plots in file Plots.ps can be examined to compare the estimated distribution
  of individual admixture proportions (based on the the posterior means for
  individual admixture) with an estimate for the distribution of individual
  admixture values in the population (based on the posterior means for the
  Dirichlet parameters of this distribution).<span style='mso-spacerun:yes'>  
  </span></span><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p><span style='mso-bidi-font-style:italic'>&nbsp;</span><a name="Contact:"><span
  style='font-size:13.5pt;font-family:Arial;color:#214772;mso-bidi-font-style:
  italic'>Contact:</span></a><span style='mso-bidi-font-style:italic'><o:p></o:p></span></p>
  <p><span style='mso-bidi-font-style:italic'><a
  href="mailto:david.odonnell@ucd.ie"><span style='font-family:Arial'>david.odonnell@ucd.ie</span></a>
  <o:p></o:p></span></p>
  <p><span style='mso-bidi-font-style:italic'><a
  href="mailto:paul.mckeigue@ucd.ie"><span style='font-family:Arial'>paul.mckeigue@ucd.ie</span></a>
  <o:p></o:p></span></p>
  <p><span style='font-family:Arial;mso-bidi-font-style:italic'>last modified: </span><st1:date
  Year="2004" Day="13" Month="12"><!--webbot bot="Timestamp"
  S-Type="EDITED" S-Format="%d %B %Y" startspan -->08 February 2005<!--webbot bot="Timestamp" endspan i-CheckSum="41129" -->
  </st1:date><span style='mso-bidi-font-style:
  italic'><o:p></o:p></span></p>
  <p><i>&nbsp;<o:p></o:p><o:p></o:p></i></p>
  </div>
  <p style='margin-left:7.15pt'><i>&nbsp;<o:p></o:p></i></p>
  <p style='margin-left:7.15pt'><i>&nbsp; <o:p></o:p></i></p>
  </td>
 </tr>
</table>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

</div>

</body>

</html>
