<html>
<head>
<title>Admixmap Tutorial</title>
</head>
<body>

<h1>Tutorial on using ADMIXMAP to model genotype and phenotype data from an admixed population</h1>
<p>If you have problems getting the program to run, email David O'Donnell <a href="mailto:nigel.wetters@lshtm.ac.uk">&lt;david.odonnell@ucd.ie&gt;</a>.  To feed back comments on this tutorial, email Paul Mckeigue <a href="mailto:paul.mckeigue@lshtm.ac.uk">&lt;paul.mckeigue@ucd.ie&gt;</a>.</p>

<h2>Getting started</h2>
<p>This tutorial is based on the dataset "philadelphia", provided as a test dataset for download, prepared by Indrani Halder and Dr Mark Shriver at Penn State.  The data are based on a sample of 202 African-Americns, and the following data files have been prepared for input to ADMIXMAP.</p>

 
<table>
<tr>
  <td>Filename</td>
  <td>Contents</td>
</tr>
<tr>
  <td>outcomevars.txt</td>
  <td>hypertension, coded as 0=unaffected, 1=affected.</td>
</tr>
<tr>
  <td>covariates.txt</td>
  <td>age, sex and body mass index, standardized about their sample means. </td>
</tr>
<tr>
  <td>genotypes.txt</td>
  <td>genotypes at 26 SNP loci</td>
</tr>
<tr>
  <td>loci.txt</td>
  <td>locus description file, with locus name, number of alleles, and map distance from last locus</td>
</tr>
<tr>
  <td>stratificationtestloci.txt</td>
  <td>list of numbers representing loci to be included in the test for stratification</td>
</tr>
<tr>
  <td>priorallelefreq2pops.txt</td>
  <td>parameters for prior distributions of ancestry-specific allele frequencies (west African and European) at each locus. Calculated by adding 0.5 to the values in historicallelefreq.txt </td>
</tr>
<tr>
  <td>priorallelefreq3pops.txt</td>
  <td>parameters for prior distributions of ancestry-specific allele frequencies (west African, European and Native American) at each locus.  Calculated by adding 0.5 to the values in historicallelefreq.txt</td>
</tr>
<tr>
  <td>historicallelefreq.txt</td>
  <td>observed counts of each allele at each locus in samples from modern unadmixed west Africans, Europeans and Native Americans</td>
</tr>
<tr>
  <td>allelefreq.txt</td>
  <td>allele frequencies at each locus in west Africans and Europeans, specified as constants</td>
</tr>
</table>

<p>These notes assume that you are working on a Linux machine with Perl, the R statistical package, and a viewer for postscript files installed.</p>

<p>Create a working directory for this exercise, and copy these data files into a subdirectory (of your working directory) named <b>data</b>. Copy the Perl script (philadelphia.pl) into your working directory.  Copy the the main ADMIXMAP program file and the R script AdmixmapOutput.R to the current working directory - if you put them somewhere else, you will have to edit the Perl script to give the correct pathnames.</p>

<p>You are now ready to start the Perl script from your working directory.  This script will run a series of analyses with different models, calling the main ADMIXMAP program each time with appropriate command-line options.  After each ADMIXMAP analysis has completed, the Perl script will clean up the output files, run the R script AdmixmapOutput.R to analyse the output files, and move all output files to a subdirectory named for the type of analysis that has been run.  Six new subdirectories, each containing files output by ADMIXMAP and the R script AdmixmapOutput.R, will be created.  On a fast PC, these analyses will take about 20 minutes.  For speed, the Perl script has been set to compute only 2500 iterations.  For serious work, we would recommend running at least ten times as many iterations.</p>

<p>We now examine the results of each ADMIXMAP analysis.</p>

<h2>SinglePopResults</h2>
<p>This directory contains results of analysis with command-line option <code>populations=1</code>.  No information about allele frequencies is supplied. The program fits a logistic regression model with hypertension as outcome variable, and age, sex and body mass index as explanatory variables.  Each locus is tested for allelic association with hypertension. </p>

<p>The file BurninTestStats.txt contains results of a simple test for the adequacy of the burn-in period, attributed to Geweke (1992).  Ignore the first line of this file when run with option <code>populations=1</code>.  If burn-in is adequate, these test statistics should have a standard normal distribution.  However this test is not very sensitive, and may fail to detect lack of convergence.</p>

<p>The file TestForStratificationCumulative.txt contains successive evaluations of a global test for residual stratification, evaluated over all posterior samples obtained since the end of the burn-in period and written to file every 100 iterations (10 times the command-line option <code>every</code>).  Only the last value is used to test for stratification - examination of successive values allows you to estimate whether the sampler has been run long enough for the test statistic to be estimated accurately.  See the main program documentation for more details of how the test statistic is calculated.  The value of the test statistic is about 0.45 (compared with an expected value of 0.5), indicating that there is not strong evidence of stratification in this dataset.   We therefore do not expect there to be strong confounding of allelic associations by population stratification in this dataset, nor do we expect to be able to estimate associations of the outcome variable with individual admixture accurately (because these estimates depend upon variation of admixture proportions between individuals).</p>

<p>Next examine the file TestsAllelicAssociationFinal.txt, which contains score tests of allelic association at each locus.  The score test for a null hypothesis that a scalar variable &beta;=0, is calculated by dividing the score (gradient of the log-likelihood at &beta;=0) by the square root of the observed information (curvature of the log-likelihood at &beta;=0).  This statistic has a standard normal distribution under the null hypothesis.  Positive score values indicate that the most likely value of &beta; is greater than zero.  The proportion of information extracted is a measure of how much information about &beta; we have, in comparison with a dataset in which all variables were observed directly.</p>

<p>The p-values in this table will be practically identical to those that would be obtained by testing for association with number of copies of each allele in a classical logistic regression analysis, adjusting for age, sex and body mass index.  The only advantage of using ADMIXMAP for this type of analysis is that where two or more loci in the same gene have been typed, ADMIXMAP will model the unobserved haplotypes and compute p-values for associations with haplotypes that correctly allow for uncertainty in haplotype assignments.  Only one locus (TYR192) shows strong evidence of association in this analysis.</p>

<p>Note that the proportion of information extracted is close to 100% at all loci - this is because with a model based on a single population, and no uncertainty about haplotype assignments, the score does not vary over the posterior distribution.</p>

<h2>TwoPopsResults</h2>
<p>This directory contains results of analysis with command-line option <code>populations=2</code>.   The model fits a logistic regression as previously, with individual admixture proportion, together with age, sex, and body mass index, as explanatory variables.  Even if your objective is only to study population structure, and not to study the outcome variable, including an outcome variable (such as skin pigmentation) that is strongly related to individual admixture proportions helps the program to learn about individual admixture proportions and population stratification.</p>

<p>The test statistic for stratification is now about 0.63 - there is now no evidence of residual stratification not accounted for by a model with two subpopulations.</p>

<p>As the regression model includes individual admixture proportion, tests of allelic association will be adjusted for population stratification is inferred by the model.  The p-value for association of hypertension with TYR-192 is still statistically significant at p&lt;0.01, from which we can infer that the crude association is unlikely to be accounted for by confounding by population stratification.  Note that the proportion of information extracted is now only about 90% at most loci, compared with 100% in the single population model - this is because individual admixture is estimated from only 26 markers, and the score statistic (which is based on adjusting for individual admixture in the regression model) consequently varies over the posterior distribution.</p>

<p>To determine whether the sampler has been run long enough to estimate the p-value accurately, we can examine the postscript file TestsAllelicAssociation.ps.  This shows the results of successive evaluations of the score test, evaluated over all posterior samples obtained since the end of the burn-in period and written to file every 100 iterations.  Even with this short run of only 2000 iterations, the p-values appear to be stable.</p>

<p>Estimates of individual admixture proportions obtained in this analysis are not meaningful because unless at least some information about allele frequencies or individual ancestry is supplied, the subpopulations are not identifiable in the model.  To rank individuals by their "degree of admixture", the posterior means of "ancestry diversity" (third column in file IndividualVarPosteriorMeans.txt) can be examined.  This is calculated as the probability that locus ancestry differs between two gene copies drawn at random from the same individual at unlinked loci.  It does not depend upon the labelling of the subpopulations, and so can be computed even when the subpopulations are not identifiable.  If you are using the program to identify admixed individuals, without using prior information about allele frequencies, we recommend that you use this statistic.</p>

<h2>FixedFreqResults</h2>
<p>This directory contains results of fitting a model with fixed allele frequencies, specified in the file <code>allelefreqfile</code>. This does not allow for any sampling error in the estimates of allele frequencies in the unadmixed modern descendants.  We can test the assumption that ancestry-specific allele frequencies have been specified correctly by examining the score tests for mis-specified allele frequencies.  Successive evaluations of these score tests are in file TestForMisSpecAlleleFreqs.txt.  The strongest evidence for mis-specified allele frequencies is at CYP19, although several other loci - MID52, WI11909, WI14319 - are statistically significant).   On this basis, we should fit a model that does not assume that allel frequencies can be specified correctly.</p>

<h2>PriorFreq2PopsResults</h2>
<p>This directory contains result of analysis with command-line option <code>priorallelefreqfile=allelelefreq2pops.txt</code>. With this option, the program fits a model in which allele frequencies in the unadmixed populations sampled are assumed to be identical to the corresponding ancestry-specific allele frequencies in the admixed population.  Specifying a prior distribution for the allele frequencies, rather than specifying them as fixed constants, allows for the uncertainty in estimates of allele frequencies that are based on samples of finite size.  At locus ASIP3, where no prior information about allele frequencies is available, a reference prior (all parameters equal to 0.5) is specified.  Where haplotype frequencies have been estimated from phase-unknown genotypes, the uncertainty in these estimates can also be accounted for in the parameters of the prior distribution.</p>

<p>The file PosteriorQuantiles.txt contains the posterior mean, posterior median and central 95% credible interval for each population-level variable: Dirichlet parameters for the distribution of admixture, all regression coefficients, and the population admixture proportions.  If the study sample size is large, the posterior mode and 95% credible interval are asymptotically equivalent to the maximum likelihood estimate and 95% confidence interval.  If the sample size is large, and the variable has been transformed to lie on the real line (between minus infinity and plus infinity), the posterior distribution will be approximately normal and thus the posterior mode will be approximately equal to the mean or median.  Plots of the posterior densities of all population-level parameters are given in the postscript file PosteriorDensities.ps.  The table below briefly explains what the variables tabulated in this file mean.</p>

<table>
<tr>
  <td>Variable</td>
  <td>Explanation</td>
</tr>
<tr>
  <td>Dirichlet parameters for distribution of admixture (Afr, Eur) in the population</td>
  <td>Large values of the Dirichlet parameters imply that the variance of individual admixture proportions is small.</td>
</tr>
<tr>
  <td>Sum of intensities</td>
  <td>This parameter measures the frequency with which transitions between states of African and European ancestry occur along the chromosomes in this population.  The current version of the program assumes this parameter to be the same in all individuals.   A large value for the sum of intensities indicates that the average number of generation back to an unadmixed African or European ancestor is large.</td>
</tr>
<tr>
  <td>Regression coefficients for intercept, covariates, and individual admixture</td>
  <td>These are logistic regression coefficients (log odds ratios) and can be transformed to odds ratios by taking exponents.  With K populations, K - 1 regression coefficients for individual admixture proportions are displayed (the last population is taken as the baseline category).</td>
</tr>
<tr>
  <td>Population admixture proportions</td>
  <td>Population admixture proportions are are calculated by dividing the Dirichlet parameters by their sum.</td>
</tr>
<tr>
  <td>EffectVsAllOther (calculated only with 3 or more subpopulations in the model)</td>
  <td>For each subpopulation, the difference between the predicted log odds of hypertension at admixture proportions of 1 and 0 for that subpopulation.  This allows you to estimate the effect of admixture proportions from one subpopulation, irrespective of the others.</td>
</tr>
</table>


<p>The postscript file AutocorrelationPlots.ps contains plots of the autocorrelation between successive draws from the posterior samples of the population-level variables.  Strong autocorrelation extending over 50 or more iterations indicates that the mixing of the sampler is poor.  Long sampling runs are then required to estimate parameters and test statistics.  Centering the covariates about their sample means helps to reduce autocorrelations.  The postscript file ParameterCumulativeMeans.ps contains plots of the ergodic averages (cumulative means) of the population-level variables.  This can be examined to determine whether the sampler has been run long enough to estimate these means accurately.</p>

<p>Posterior means of individual admixture are in the file IndividualVarPosteriorMeans.txt.  These values can be plugged into other types of genetic analysis, but should not be used to test for a relationship with the outcome variable because any association between individual admixture and the outcome variable has already been used by the program to learn about individual admixture.  If you want estimates of individual admixture that you can plug into a regression model to test for association with the outcome variable, run the program with option <code>analysistypeindicator=1</code>, which does not fit a regression model.</p>

<p>The file DistributionIndividualAdmixture.ps contains histograms of the distribution of posterior means of individual admixture.  For comparison, the distribution of individual admixture specified by the posterior means of the Dirichlet parameters is shown as a curve on the same plot.  This file can be examined to test if there is any obvious lack of fit of the distribution of individual admixture proportions to the model: for instance a bimodal distribution of admixture proportions, which is not compatible with a Dirichlet distribution.</p>

<p>The file TestsForDispersion.txt contains cumulativeresults of a test for dispersion (variation) of allele frequencies between the unadmixed populations (in this example west African and European) that were sampled to obtain prior parameters in <code>priorallelefreqfile</code>, and the corresponding ancestry-specific allele frequencies in the admixed population (in this case African-American).  Ignore the zero value at ASIP3 - the test statistic is not meaningful at loci where a reference prior is given in <code>priorallelefreqfile</code>.  Also ignore the global "Population" test statistic which is evaluated over all loci including ASIP3.  This will be fixed in the next release of the ADMIXMAP program.</p> 

<p>At most loci, the test statistic is close to 0.5, except for CYP19 where the test statistic for west Africans is only 0.04.  As "Bayesian p-values" are conservative in comparison with classical p-values, this can be interpreted as strong evidence that allele frequencies at this locus vary between the west African populations sampled and the pool of genes of African ancestry within the African-American population of Philadelphia.  One way to deal with this in further analysis of the data would be to replace the prior for allele frequencies at CYP19 by a reference prior, allowing the program to learn about allele frequencies from the admixed population without using any prior information.  Alternatively, where several loci show evidence of dispersion of allele frequencies, we can fit a dispersion model as described in the next section.</p>

<p>The results of tests of association of hypertension with alleles at each locus are similar to those obtained in the TwoPopsResults analysis, in which no prior information about allele frequencies was supplied.</p>

<p>In this analysis, where the subpopulations are identifiable, we can test also for association of the outcome variable with ancestry at each locus.  This is the most appropriate analysis if your objective is to exploit admixture to localize genes underlying ethnic variation in disease risk, rather than to test a candidate gene for association with the outcome variable while adjusting for individual admixture proportions as a possible confounder.  The results of these tests are in the file TestsAncestryAssociationFinal.txt, with plots of successive evaluations in the postscript file TestsAncestryAssociation.ps.  Note that the efficiency of this test (proportion of information extracted) is much lower than the efficiency of the test for allelic association: the proportion of information extracted ranges from 81% at locus FYNULL, which is almost completely informative for west African versus European ancestry, to near zero at loci that are uninformative for west African versus European ancestry, such as CKMM.  At some loci the information is negative - probably because the sampler has not been run long enough, although it is formally possible for the information to be negative (log-likelihood not concave downwards) in small samples.  There is evidence of linkage to genes underlying the ethnic differences in disease risk at TYR192, and some suggestion of an effect at D11S429, which lies 13 cM to the left of TYR192.   The test for linkage with ancestry, unlike the test for allelic association, uses the genotype data from all loci on this chromosome to extract information about ancestry at each locus.  With this test, we expect any evidence of linkage to be detectable over a broad region: 20 cM or more.  The score for African ancestry is positive, implying that the risk of hypertension increases with the proportion of gene copies that are of African ancestry at this locus.  To establish whether this finding can be confirmed, the next step would be to type more markers informative for west African versus European ancestry in this region.</p>

<p>The file TestsAncestryAssociation.ps also contains a plot of the proportion of information extracted at each locus, across all loci.  With larger marker sets, this can be used to evaluate the coverage of each chromosome by the marker set.</p>

<h2>Dispersion2PopsResults</h2>
<p>This directory contains results of analysis with command-line option <code>historicallelefreqfile=historicallelefreq2pops.txt</code>.  With this option, the program fits a "dispersion" model for the allele frequencies, which allows the allele frequencies in the unadmixed populations to vary from the corresponding ancestry-specific allele frequencies in the admixed population under study.  A single dispersion parameter (eta) is estimated for each subpopulation.  The results of these analyses should be interpreted with caution for the present, as we have only recently added this capability to the program and our experience with this type of model is limited.</p>

<p>This dispersion parameter can be interpreted as follows.  Imagine that the variation between allele frequencies in the modern unadmixed west African populations sampled and the allele frequencies in the pool of genes of African ancestry in the African-American population of Philadelphia had been generated by drawing two independent equal-sized samples from an ancestral total population.  The allele frequencies in the two samples would differ as a results of sampling variation, and the variance of the sample frequencies between these two samples would depend upon the size of the sample that was drawn.  The parameter eta is the sample size that accounts for this variation.  Small values of eta (&lt;100) indicate dispersion of allele frequencies.  Eta is related to Wright's F<sub>ST</sub> by the relation F<sub>ST</sub>=1/(1+eta).</p>

<p>The posterior distributions of the allele frequency dispersion parameters (labelled as eta.Afr and eta.Eur for African and European subpopulations respectively) are plotted in the file ParameterPosteriorDensities.ps, and the means and medians are given in the file PosteriorQuantiles.txt</p>

<p>The results of this analysis suggest that there is more dispersion of allele frequencies between the unadmixed populations sampled and the corresponding ancestry-specific allele frequencies in the African-American population is greater for west Africans than for Europeans.  This is to be expected, as there is greater genetic heterogeneity between west African subpopulations than between European subpopulations, and we are unable to sample the exact mix of west African subpopulations that contributed genes to the modern African-American population.  This has implications for the sample size required when estimating allele frequencies in unadmixed populations in order to model admixture.  In general, there is no point in using a sample size that is an order of magnitude larger than the value of eta for a given subpopulation, because the allele frequencies in the unadmixed populations will not accurately predict the corresponding ancestry-specific allele frequencies in the admixed population.</p>

<p>The results of tests for allelic association and linkage with ancestry are similar to those obtained with the previous analysis using a model with no dispersion, suggesting that these analyses are fairly robust to assumptions about allele frequencies.</p>

</body>
